---
title: "Batch effect correction"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = FALSE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "right",  buttonSide = "right")`

# Getting ready

## Load packages

```{r}
library(here) # A Simpler Way to Find Your Files, CRAN 
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN  
library(qiime2R) # Import qiime2 artifacts to R, [github::jbisanz/qiime2R]   
library(mixOmics) # Omics Data Integration, Bioconductor  
#library(RUVSeq) # Remove Unwanted Variation from RNA-Seq Data, Bioconductor 
#library(sva) # Surrogate Variable Analysis, Bioconductor  
library(patchwork) # Streamlined Plot Theme and Plot Annotations for 'ggplot2', CRAN   
library(vegan) # Community Ecology Package, CRAN  
```

## Import data

```{r}
# metadata
mtd <- read_tsv(here("data/metadata.tsv"), comment = "#q2")

# feature table
asv_tbl <- read_qza(here("data/intermediate/qiime2/asv/table-filtered-sepp-inserted.qza")) %>%
  pluck("data") 
```

## Data preprocessing

```{r}
# metadata
mtd <- mtd %>% 
  filter(!SampleType %in% c("Extraction-blank", "PCR-blank", "Mock")) %>%
  rename(SampleID = "#SampleID") %>% 
  column_to_rownames("SampleID") %>% 
  mutate(
    PCRBatch = factor(PCRBatch),
    SampleType = factor(
      SampleType, 
      levels = c("REF-PID", "REF-PIM", "REF-DID", "REF-DIM", "IM-PID", "IM-PIM", 
                 "IM-DID", "IM-DIM", "Water", "Feed")
      )
    )  

# feature table
asv_tbl <- asv_tbl[, rownames(mtd)]
```

# Assess batch effect

```{r}
# Centered log-ratio transformation
asv_tbl_clr <- logratio.transfo(t(asv_tbl), logratio = "CLR", offset = 1 ) %>% 
  as("matrix")

# PCA
pca_before <- pca(asv_tbl_clr, ncomp = 3)

# PCA plot
pcaPlot_before <- cbind(mtd, pca_before$variates$X) %>%
  ggplot(aes(x = PC1, y = PC2, color = SampleType, shape = PCRBatch)) +
  geom_point(size = 2) +
  # connect paired technical replicates by line
  geom_line(aes(group = SampleName), color = "black") +
  labs(
    title = "Before batch correction", color = "Sample type", shape = "PCR batch", 
    x = paste0("PC1: ", round(100 * pca_before$explained_variance[1], 2), "%"),
    y = paste0("PC2: ", round(100 * pca_before$explained_variance[2], 2), "%")
    ) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() 

pcaPlot_before 
```

From the PCA plot, we can tell that technical replicates (connected by the lines) are far apart. Mucosa samples (in dark green and blue color) from fish fed the REF diet form two clusters based on PCR batches. 

# Correct batch effect

## RUVSeq

[RUVSeq](https://www.nature.com/articles/nbt.2931) removes unwanted variation in RNASeq data using replicate samples and negative control genes. Here we use one of its commonly used methods, RUVs, for correcting batch effects in microbiome data. The two main parameters of RUVs are the number of unwanted factors, k, and the set of negative control variables genes, or ASVs in our case. The choice of the parameter k is not easy and is data-dependent. Empirically, a few (2-3) factors are enough to capture the unwanted variation. In noisy data, K can be increased to 5 or 10. Note that if k is too high, the RUVs over-corrects for unwanted variation and ends up removing (almost) all the biological variability within the conditions. The choice of negative control variables is also somewhat data-dependent. However, RUVs is robust to the choice of negative control variables. One is recommend to perform extensive exploratory data analysis, comparing different values of k and sets of negative control variables. 

```{r}
# sort ASVs by variance across samples
asv_tbl <- as.data.frame(asv_tbl) %>%
  rownames_to_column("featureID") %>%
  mutate(
    sd   = apply(.[, names(.) != "featureID"], 1, sd),
    mean = apply(.[, names(.) != "featureID"], 1, mean),
    cv   = sd / mean
    ) %>%
  arrange(cv) %>% # arrange ASVs by their CV
  dplyr::select(-c(sd:cv)) %>%
  column_to_rownames("featureID") %>%
  as.matrix()

# define the proportions of ASVs to be used as negative controls
prp <- c(0.1, 0.2, 0.5, 1) 

# define the number of unwanted variation factors 
k <- c(1, 2, 3, 4, 5, 10) 

# get combinations of k and prp for the parameter tuning
prmt <- expand.grid(prp = prp, k = k)

# sample replicate matrix
SampleType <- mtd$SampleType
names(SampleType) <- rownames(mtd)
rpl <- RUVSeq::makeGroups(SampleType)

# batch effect correction
ruvs <- map2(
  prmt$prp,
  prmt$k,
  function(prp, k) 
  {
  # number of negative control ASVs 
  nmb_nc_asv <- round(nrow(asv_tbl) * prp)
  
  # logical vector indicating which ASVs to be used as negative controls
  nc_vct <- rep(c(TRUE, FALSE), times = c(nmb_nc_asv , nrow(asv_tbl) - nmb_nc_asv))
    
  # run RUVs
  ruvs <- RUVSeq::RUVs(
    x = asv_tbl, # raw count matrix
    cIdx = nc_vct, # negative control features
    k = k, # the number of unwanted factors 
    scIdx = rpl # sample replicate matrix
    )
  }
)

# assign names to list elements
names(ruvs) <- paste0("ruvs_k", prmt$k, "_nc", prmt$prp)
```

## ComBat-Seq

[ComBat-seq](https://academic.oup.com/nargab/article/2/3/lqaa078/5909519) is a batch effect correction method for RNASeq data. It is an improved model based on the popular effect correction tool, [ComBat](https://academic.oup.com/biostatistics/article/8/1/118/252073).ComBat-seq takes raw count matrix as input. Same as ComBat, it requires a known batch variable.

```{r}
# PCR batch
PCRBatch <- mtd$PCRBatch
names(PCRBatch) <- rownames(mtd)

# model matrix
mod_mtrx <- model.matrix(~SampleType)

# run CombatSeq
cmbt <- sva::ComBat_seq(asv_tbl, batch = PCRBatch, covar_mod = mod_mtrx) 
```

# Evaluate methods

## Visual inspection: PCA plot

### RUVSeq

Extract and log ratio transform batch effect corrected ASV table.

```{r}
ruvs_clr <- purrr::map(ruvs, ~pluck(.x, "normalizedCounts")) %>% # get batch corrected count table
  purrr::map(~t(.x)) %>% # transpose normalized count table
  purrr::map(~logratio.transfo(.x, logratio = 'CLR', offset = 1)) %>% # CLR transformation
  purrr::map(~as(.x, "matrix")) 
```

Run PCA.

```{r}
pca_ruvs <- purrr::map(ruvs_clr, ~pca(.x, ncomp = 3)) 
```

Make PCA plots.

```{r, fig.width=16, fig.height=20}
# make plots
pcaPlot_ruvs <- pmap(list(pca_ruvs, prmt$k, prmt$prp), function(x, y, z){
  cbind(mtd, x$variates$X) %>%
  ggplot(aes(x = PC1, y = PC2, color = SampleType, shape = PCRBatch)) +
    geom_point(size = 2) +
    # connect paired technical replicates by line
    geom_line(aes(group = SampleName), color = "black") +
    labs(
      title = paste0("RUVs (k = ", y, ", nc = ", z, ")"), 
      color = "Sample type", shape = "PCR batch", 
      x = paste0("PC1: ", round(100 * x$prop_expl_var$X[1], 2), "%"),
      y = paste0("PC2: ", round(100 * x$prop_expl_var$X[2], 2), "%")
      ) +
    scale_color_brewer(palette = "Paired") +
    theme_bw() 
    }
  ) 

# arrange plots
wrap_plots(pcaPlot_ruvs, ncol = 4, guides = "collect") 
```

### ComBat-Seq

```{r}
# centered log-ratio transformation
cmbt_clr <- logratio.transfo(t(cmbt), logratio = 'CLR', offset = 1) %>% 
  as("matrix")

# PCA
pca_cmbt <- pca(cmbt_clr, ncomp = 3)

# make PCA plot
pcaPlot_cmbt <- cbind(mtd, pca_cmbt$variates$X) %>%
  ggplot(aes(x = PC1, y = PC2, color = SampleType, shape = PCRBatch)) +
  geom_point(size = 2) +
  geom_line(aes(group = SampleName), color = "black") +
  labs(title = "ComBat-Seq", 
       color = "Sample type", shape = "PCR batch", 
       x = paste0("PC1: ", round(100 * pca_cmbt$prop_expl_var$X[1], 2), "%"),
       y = paste0("PC2: ", round(100 * pca_cmbt$prop_expl_var$X[2], 2), "%")) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() 

pcaPlot_cmbt
```

## Variance explained

### Run pRDA

The partial redundancy analysis (pRDA) is a multivariate method to assess globally the effect of treatments and batch. Here, we run pRDA and get variance explained by treatments and batch.

```{r}
# combine ASV table before and after batch corrections in a list
asv_tbls <- c(before = list(asv_tbl_clr), ruvs_clr, combatseq = list(cmbt_clr)) 

# merge ASV tables and nest by batch correction methods 
asv_tbls_nst <- purrr::map(asv_tbls, as.data.frame) %>%
  bind_rows(.id = "batch_correction") %>%
  group_nest(batch_correction, .key = "data")

# define experiment design
design <- data.frame(
  SampleType = rep(SampleType, length(asv_tbls)), 
  PCRBatch = rep(PCRBatch, length(asv_tbls)), 
  batch_correction = rep(names(asv_tbls), each = nrow(mtd))
  )

design_nst <- group_nest(design, batch_correction, .key = "design")

# run pRDA and extract variance explained by treatment and batch effects
rda <- inner_join(asv_tbls_nst, design_nst, by = "batch_correction") %>%
  mutate(
    rda_trt = map2(data, design, ~rda(.x ~ PCRBatch + Condition(SampleType), data = .y)), 
    rda_bat = map2(data, design, ~rda(.x ~ SampleType + Condition(PCRBatch), data = .y)),
    var_trt = map_dbl(rda_trt, ~.x$pCCA$tot.chi*100/.x$tot.chi),
    var_bat = map_dbl(rda_bat, ~.x$pCCA$tot.chi*100/.x$tot.chi),
    batch_correction = factor(batch_correction, levels = names(asv_tbls))
    ) %>%
  dplyr::select(batch_correction, var_trt, var_bat)  %>%
  rename(Method = batch_correction, Treatment = var_trt, Batch = var_bat)  %>%
  pivot_longer(Treatment:Batch, names_to = "Type", values_to = "var_expl")  %>%
  mutate(var_expl = round(var_expl, 2))
```

### Plotting

Make bar plots showing variance explained by treatments and batch effects.

```{r, fig.height=8, fig.width=9}
var_expl <- ggplot(rda, aes(x = fct_rev(Method), y = var_expl, fill = Type)) + 
  geom_bar(stat = "identity", position = 'dodge', colour = 'black') + 
  geom_text(
    aes(Method, var_expl + 3, label = var_expl), size = 3,
    position = position_dodge(width = 0.9)
    ) + 
  coord_flip() +
  scale_y_continuous(limits = c(0, 100), expand = expansion(mult = c(0, 0))) +
  labs(x = "", y = "Variance explained (%)") + 
  theme_bw(base_size = 14) +
  guides(fill = guide_legend(reverse = TRUE)) 

var_expl
```

# Summary

Based on the PCA plots and partial redundancy analysis (pRDA), we can conclude that the RUVs provides better batch effect correction than the Combat-Seq does. In agreement with previous observations, the RUVs is robust to the choice of negative control variables. Increasing the number of unwanted factors, k, removes more unwanted variation. Setting k = 4, 5 or 10 yields very good and similar results. However, our downstream data analyses show that some results are contradictory before and after the batch effect correction (e.g., microbial overlap between feed and intestinal mucosa; alpha-diversity). Hence, we'll not correct the batch effect but rather analyze the data separately or account for the batch effect in the statistical models where applicable.

# Acknowledgements

We’d like to thank Wang and Lê Cao for publishing the code and tutorials along with their review paper ["Managing batch effects in microbiome data"](https://academic.oup.com/bib/article-abstract/21/6/1954/5643537?redirectedFrom=fulltext). Part of their code is used in the present analysis.

# Session information

```{r session-info}
sessionInfo()
```
