---
title: "Import QIIME2 artifacts into R"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = FALSE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "left",  buttonSide = "left")`

# Getting ready

Set system locale

```{r}
Sys.setlocale(category = "LC_ALL", locale = "Greek")
```

Load packages

```{r load-package}
library(here) # A Simpler Way to Find Your Files, CRAN v1.0.1
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN v1.3.0 
library(qiime2R) # Import qiime2 artifacts to R, [github::jbisanz/qiime2R] v0.99.35  
library(phyloseq) # Handling and analysis of high-throughput microbiome census data, Bioconductor v1.34.0 
```

# Phyloseq object

## Metadata

```{r}
mtd <- read_tsv(here("data/metadata.tsv"), comment = "#q2") %>%
  rename(SampleID = "#SampleID") %>% 
  column_to_rownames("SampleID") %>% 
  mutate(Diet = factor(Diet, levels = c("REF", "IM")),
         Compartment = factor(Compartment, levels = c("PI", "DI")),
         SampleOrigin = factor(
           SampleOrigin, 
           levels = c("Feed", "Water", "Digesta", "Mucosa", 
                      "Mock", "DNA-extraction", "Amplicon-PCR")),
         SampleType = factor(
           SampleType, 
           levels = c("REF-PID", "REF-PIM", "REF-DID", "REF-DIM", 
                      "IM-PID", "IM-PIM", "IM-DID", "IM-DIM", 
                      "Feed", "Water", "Extraction-blank", "PCR-blank", "Mock")),
         PCRBatch = factor(PCRBatch, levels = 1:3)) 
```

## Feature table

```{r}
otu <- read_qza(here("data/intermediate/qiime2/97otu/table-filtered-97otu-sepp-inserted.qza")) %>%
  pluck("data") %>% 
  as("matrix")
```

## Taxonomy

```{r}
txnm <- read_qza(here("data/intermediate/qiime2/asv/taxonomy-silva132.qza"))
txnm <- txnm$data %>% 
  as.data.frame() %>%
  mutate(Taxon = gsub("D_0", "k", Taxon), Taxon = gsub("D_1", "p", Taxon),
         Taxon = gsub("D_2", "c", Taxon), Taxon = gsub("D_3", "o", Taxon),
         Taxon = gsub("D_4", "f", Taxon), Taxon = gsub("D_5", "g", Taxon),
         Taxon = gsub("D_6", "s", Taxon)) %>%
  separate(Taxon, sep = ";", c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>% 
  column_to_rownames("Feature.ID") %>%
  select(-Confidence)
```

## Phylogeny

```{r}
tree <- read_qza(here("data/intermediate/qiime2/97otu/insertion-tree-97otu.qza"))
```

## Assemble a phyloseq object

Assemble data.

```{r}
ps <- phyloseq(sample_data(mtd), 
               otu_table(otu, taxa_are_rows = TRUE),
               tax_table(as.matrix(txnm)),
               phy_tree(tree$data))
```

# Alpha-diversity

```{r}
# get absolute file path for alpha diversity indices
fp_alph <- list.files(here("data/intermediate/qiime2/97otu/core-metrics-results/"), 
                      pattern = "*vector.qza", full.names = TRUE)

# get alpha diversity index names
names_alph <- sapply(strsplit(basename(fp_alph), "_vector"), `[`, 1)

# import alpha diversity indices
alph <- lapply(fp_alph, read_qza)
names(alph) <- names_alph
```

# Beta-diversity

## Distance metrics

### Digesta samples

```{r}
# get absolute file path of distance metrics
fp_dist_dgst <- list.files(here("data/intermediate/qiime2/97otu/core-metrics-results-digesta/"), 
                           pattern = "*distance_matrix.qza", full.names = TRUE)

# get distance metric names
names_dist_dgst <- sapply(strsplit(basename(fp_dist_dgst), "_distance"), `[`, 1)

# import distance metrics
dist_dgst <- lapply(fp_dist_dgst, read_qza)
names(dist_dgst) <- names_dist_dgst

# import Aitchison distance
dist_dgst_atch <- read_qza(here("data/intermediate/qiime2/97otu/robust-Aitchison-pca-digesta/distance_matrix.qza"))

# add Aitchison distance to the distance metric list
dist_dgst <- c(dist_dgst, aitchison = list(dist_dgst_atch))
```

### Mucosa, feed and water samples

```{r}
# get absolute file path of distance metrics
fp_dist_mcsFdWt <- list.files(here("data/intermediate/qiime2/97otu/core-metrics-results-mucosa-feed-water/"), 
                              pattern = "*distance_matrix.qza", full.names = TRUE)

# get file names
names_dist_mcsFdWt <- sapply(strsplit(basename(fp_dist_mcsFdWt), "_distance"), `[`, 1)

# import distance metrics
dist_mcsFdWt <- lapply(fp_dist_mcsFdWt, read_qza)
names(dist_mcsFdWt) <- names_dist_mcsFdWt

# import Aitchison distance
dist_mcsFdWt_atch <- read_qza(here("data/intermediate/qiime2/97otu/robust-Aitchison-pca-mucosa-feed-water/distance_matrix.qza"))

# add Aitchison distance to the distance metric list
dist_mcsFdWt <- c(dist_mcsFdWt, aitchison = list(dist_mcsFdWt_atch))
```

## Ordination

### Digesta samples

```{r}
# get absolute file path of pcoa results
fp_ord_dgst <- list.files(here("data/intermediate/qiime2/97otu/core-metrics-results-digesta/"), 
                          pattern = "*pcoa_results.qza", full.names = TRUE)

# get file names
names_ord_dgst <- sapply(strsplit(basename(fp_ord_dgst), "_pcoa"), `[`, 1)

# import pcoa results
ord_dgst <- lapply(fp_ord_dgst, read_qza)
names(ord_dgst) <- names_ord_dgst

# import pcoa result of Aitchison distance
ord_dgst_atch <- read_qza(here("data/intermediate/qiime2/97otu/robust-Aitchison-pca-digesta/biplot.qza"))

# add pcoa result of Aitchison distance to the pcoa result list
ord_dgst <- c(ord_dgst, aitchison = list(ord_dgst_atch))
```

### Mucosa, feed and water samples

```{r}
# get absolute file path of pcoa results
fp_ord_mcsFdWt <- list.files(here("data/intermediate/qiime2/97otu/core-metrics-results-mucosa-feed-water/"), 
                             pattern = "*pcoa_results.qza", full.names = TRUE)

# get file names
names_ord_mcsFdWt <- sapply(strsplit(basename(fp_ord_mcsFdWt), "_pcoa"), `[`, 1)

# import pcoa results
ord_mcsFdWt <- lapply(fp_ord_mcsFdWt, read_qza)
names(ord_mcsFdWt) <- names_ord_mcsFdWt

# import pcoa result of Aitchison distance
ord_mcsFdWt_atch <- read_qza(here("data/intermediate/qiime2/97otu/robust-Aitchison-pca-mucosa-feed-water/biplot.qza"))

# add pcoa result of Aitchison distance to the pcoa result list
ord_mcsFdWt <- c(ord_mcsFdWt, aitchison = list(ord_mcsFdWt_atch))
```

# Export data

```{r}
# Phyloseq object
saveRDS(ps, file = here("data/intermediate/qiime2R/phyloseq.rds"))

# Alpha-diversity indices
saveRDS(alph, file = here("data/intermediate/qiime2R/alpha_diversity.rds"))

# Beta-diversity: distance metrics
saveRDS(dist_dgst, file = here("data/intermediate/qiime2R/beta_diversity_distance_digesta.rds"))
saveRDS(dist_mcsFdWt, file = here("data/intermediate/qiime2R/beta_diversity_distance_mucosa_feed_water.rds"))

# Beta-diversity: ordinations
saveRDS(ord_dgst, file = here("data/intermediate/qiime2R/beta_diversity_ordination_digesta.rds"))
saveRDS(ord_mcsFdWt, file = here("data/intermediate/qiime2R/beta_diversity_ordination_mucosa_feed_water.rds"))
```

# Session information

```{r session-info}
sessionInfo()
```
