---
title: "Taxonomic analysis"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = FALSE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "left",  buttonSide = "left")`

# Getting ready

Load packages.

```{r load-package}
library(here) # A Simpler Way to Find Your Files, CRAN v1.0.1
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN v1.3.0 
library(cowplot) # Streamlined Plot Theme and Plot Annotations for 'ggplot2', CRAN v1.1.1  
library(ggh4x) # Hacks for 'ggplot2', CRAN v0.1.2.1 
library(ggtext) # Improved Text Rendering Support for 'ggplot2', CRAN v0.1.1
library(PerformanceAnalytics) # Econometric Tools for Performance and Risk Analysis, CRAN v2.0.4
library(RColorBrewer) # ColorBrewer Palettes, CRAN v1.1-2 
library(qiime2R) # Import qiime2 artifacts to R, [github::jbisanz/qiime2R] v0.99.35  
library(MicrobeR) # Handy functions for microbiome analysis in R, [github::jbisanz/MicrobeR] v0.3.2 
library(microbiome) # Microbiome Analytics, Bioconductor v1.12.0
library(speedyseq) # Faster implementations of phyloseq functions, [github::mikemc/speedyseq] v0.5.3.9001  #
library(flextable) # Functions for Tabular Reporting, CRAN v0.6.1    
library(officer) # Manipulation of Microsoft Word and PowerPoint Documents, CRAN v0.3.16     
library(DT) # A Wrapper of the JavaScript Library 'DataTables', CRAN v0.17 
```

Load data.

```{r}
# load phyloseq object
ps <- readRDS(here("data/intermediate/qiime2R/phyloseq.rds"))

# filter samples
ps <- subset_samples(ps, !SampleType %in% c("Extraction-blank", "PCR-blank", "Mock"))

# change OTU names for easy display
indx <- formatC(1:ntaxa(ps), width = nchar(ntaxa(ps)), format = "d", flag = "0")
taxa_names(ps) <- paste0("OTU", indx)

# extract otu table, taxonomy and metadata
otu <- as.data.frame(otu_table(ps)) 
txnm <- tax_table(ps) %>% as("matrix") %>% as.data.frame()
mtd <- data.frame(sample_data(ps), check.names = FALSE)
```

# Overview of taxonomy assignments

First of all, let's look at the coverage of taxonomy assignments at different levels.

```{r}
txnm_assg <- txnm %>%
  gather("Rank", "Name", rank_names(ps)) %>%
  group_by(Rank) %>%
  # Empty taxonomic ranks may be na or strings containing "uncultured" or "Ambiguous_taxa"
  summarize(OTUs_classified = sum(!is.na(Name) & !grepl("uncultured|Ambiguous|metagenome", Name))) %>%
  mutate(Frac_classified = OTUs_classified / ntaxa(ps),
         Frac_classified = ifelse(Frac_classified == 1, "100", round(Frac_classified * 100, 1)),
         Frac_classified = paste(Frac_classified, "%"),
         Rank = factor(Rank, rank_names(ps))) %>%
  arrange(Rank) 

txnm_assg %>%
  datatable(options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:3)))))
```

We can tell that the majority of OTUs were assigned at the genus level whereas only `r txnm_assg[7, 3]` of OTUs got a species-level annotation.

# Taxonomic composition of mock samples

Import and tidy expected mock composition

```{r}
mock_exp <- read_tsv(here("data/reference/mock_expected.tsv")) 

mock_exp <- rename(mock_exp[, 1:2], tax_exp = 1, abnd_exp = 2) %>%
  arrange(tax_exp) %>%
  mutate(tax_exp = gsub("D_0.*D_6__", "", tax_exp)) # prune taxonomy to the lowest level
```

Import and tidy observed mock composition

```{r}
mock_obs <- read_qza(here("data/intermediate/qiime2/97otu/quality-control/mock-observed-l7-rel.qza"))
mock_obs <- mock_obs$data %>%
  as.data.frame() %>%
  rownames_to_column("tax_obs") %>%
  # prune taxonomy to the lowest level
  mutate(tax_obs = gsub("D_0.*D_6__", "", tax_obs), # species level  
         tax_obs = gsub("D_0.*D_5__|;__", "", tax_obs)) # genus level
```

Merge expected and observed mock composition

```{r}
mock <- bind_cols(mock_exp, mock_obs) %>%
  mutate(gram_staining = c("(G+)", "(G+)", "(G+)", "(G+)", "(G+)", "(G-)", "(G-)", "(G-)"),
         taxa = paste0("*", tax_exp, "*", " ", gram_staining, " / ", "*", tax_obs, "*")) %>% 
  select(-tax_exp, -gram_staining, -tax_obs) %>%
  mutate(taxa = factor(taxa, levels = unique(taxa))) %>% 
  # use percentage as unit for relative abundance
  mutate_if(is.numeric, ~(100 * .x)) %>% 
  mutate_if(is.numeric, ~round(.x, 1))
```

Correlation plot

```{r}
chart.Correlation(mock[, !(colnames(mock) == "taxa")], histogram = FALSE, pch = 19)
```

Stacked barplot

```{r, fig.width=8, fig.height=5}
# plot
bar_mock <- mock %>% 
  gather(key = "type", value = "abundance", -taxa) %>%
  ggplot(aes(x = type, y = abundance, fill = taxa, label = abundance)) +
  geom_bar(stat = "identity") +
  geom_text(position = position_stack(vjust = 0.5), size = 4) +
  annotate("segment", x = 1.75, xend = 6.25, y = 102, yend = 102) +
  annotate("text", x = c(1, 4.5), y = c(105, 105), label = c("Expected", "Observed"), size = 4) +
  scale_y_continuous(limits = c(0, 105), breaks = 0:10*10, expand = expansion(mult = c(0, 0.05))) +
  labs(x = "Mock samples", y = "Relative abundance (%)", fill = "Taxa (expected (Gram stain) / observed)") +
  scale_fill_manual(values = brewer.pal(8, "Paired"), labels = levels(mock$taxa)) +
  theme_minimal() +
  theme(axis.text.x = element_blank(),
        legend.text = element_markdown()) # align legend text to left

bar_mock

# export figure
ggsave(here("result/figure/Figure S1.tiff"), width = 8, height = 5, 
       units = "in", dpi = 300, compression = "lzw")
```

# Taxonomic composition of biological samples

## Stacked bar plots

Make a table containing top 10 most abundant taxa at genus level.

```{r}
taxa_tab <- Summarize.Taxa(otu, txnm)$Genus %>% Make.Percent() 
taxa_tab <- taxa_tab[order(rowMeans(taxa_tab), decreasing = T), ]
Others <- colSums(taxa_tab[11:nrow(taxa_tab), ])
taxa_tab <- rbind(taxa_tab[1:10, ], Others)
```

Tidy dataframe for making stacked bar plots.

```{r}
taxa_tab <- as.data.frame(taxa_tab) %>%
  rownames_to_column("Taxa") %>%
  separate(Taxa, sep = ";", c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")) %>% 
  mutate(Phylum = ifelse(is.na(Phylum)|Phylum == "NA"|grepl("uncultured|Ambiguous|metagenome", Phylum), 
                         Kingdom, 
                         Phylum),
         Class = ifelse(is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
                        Phylum, 
                        Class),
         Order = ifelse(is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
                        Class, 
                        Order),
         Family = ifelse(is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
                         Order, 
                         Family),
         Genus = ifelse(is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
                        Family, 
                        Genus)) %>%
  select(-Kingdom, -(Class:Family)) %>%
  mutate(Phylum = gsub("p__", "", Phylum),
         Phylum = factor(Phylum, levels = rev(unique(Phylum))),
         Genus = gsub("g__", "", Genus),
         Genus = factor(Genus, levels = rev(unique(Genus)))) %>%
  arrange(Phylum, Genus) %>%
  mutate(Genus = factor(Genus, levels = unique(Genus))) %>%
  pivot_longer(-c(Phylum, Genus), names_to = "SampleID", values_to = "Abundance") %>%
  inner_join(rownames_to_column(mtd, "SampleID"), by = "SampleID")
```

Make stacked bar plots.

```{r, fig.width=10}
# feed samples
taxa_bar_feed <- filter(taxa_tab, SampleType == "Feed") %>%
  mutate(Diet = ifelse(SampleName == "FeedREF", "REF", "IM"),
         Diet = factor(Diet, levels = c("REF", "IM"))) %>%
  ggplot(aes(x = SampleName, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity", width = 0.5) +
  labs(x = "Sample", y = "") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_nested(~ SampleOrigin + Diet, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# water samples
taxa_bar_water <- filter(taxa_tab, SampleType == "Water") %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_wrap(~ SampleOrigin) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# samples from fish fed the REF diet 
taxa_bar_ref <- filter(taxa_tab, Diet == "REF") %>%
  # remove technical replicates
  filter(!(IsTechnicalReplicate == "yes" & PCRBatch == 2)) %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "Relative abundance (%)") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_nested(~ Diet + SampleOrigin + Compartment, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# samples from fish fed the IM diet 
taxa_bar_im <- filter(taxa_tab, Diet == "IM") %>%
  filter(!(IsTechnicalReplicate == "yes" & PCRBatch == 2)) %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_nested(~ Diet + SampleOrigin + Compartment, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# assemble plots
taxa_bar <- plot_grid(
  taxa_bar_ref, 
  taxa_bar_feed + theme(plot.margin = margin(l = -0.5, unit = "cm")),
  taxa_bar_im + theme(plot.margin = margin(l = -0.4, unit = "cm")),
  taxa_bar_water + theme(plot.margin = margin(l = -0.3, unit = "cm")), 
  nrow =  1, align = 'h', axis = "tb", 
  rel_widths = c(7.5, 1.6, 7.5, 1.4)) 

taxa_bar 
```

## Customize figure legend

Make custom legend using flextable.

```{r}
lgnd <- unique(taxa_tab[ , c("Genus", "Phylum")]) %>%
  #arrange(factor(Genus, levels = levels(taxa_tab$Genus))) %>%
  mutate(col = 10, .before = Genus) %>% # legend color
  flextable() %>% 
  theme_box() %>%
  highlight(j = "col", color = col, part = "body") %>% # use legend color to highlight text
  color(j = "col", color = col, part = "body") %>% # use legend color as font color to hide the text
  color(j = "col", color = "white", part = "header") %>% # head text in the header
  italic(i = 2:11, italic = TRUE, part = "body") %>% # italicize taxa names
  fontsize(size = 20, part = "all") %>% # font size
  width(j = c("col"), width = 0.2) %>% # change the width of legend
  height_all(height = 1) %>% # cell height
  align(j = c("Genus", "Phylum"), align = "left", part = "all") %>% # align text to the left
  merge_v(j = "Phylum") %>% # merge cells belong to the same Phylum
  border_remove() %>% # remove borders
  theme_booktabs() %>% # apply table style
  hline(i = c(1, 2), border = fp_border(width = 2), part = "body") %>% # draw horizontal lines
  hline_top(part = "all", border = fp_border(color = "white")) # remove horizontal line in the table header

# convert flextable into a ggplot object
lgnd <- as_raster(lgnd)

lgnd <- ggplot() + 
  theme_void() + 
  annotation_custom(grid::rasterGrob(lgnd, y = unit(0.5, "npc"))) # top right: y = unit(0.72, "npc")
```

## Add custom legend

Add custom legend to the plot.

```{r, fig.width=10}
taxa_bar <- plot_grid(taxa_bar, lgnd, nrow =  1, align = 'h', rel_widths = c(4, 1)) 
taxa_bar
```

# Microbial overlap

## Compute microbial overlap

Compute OTU prevalence in each sample type.

```{r}
prvl_ls <- list()

for (i in levels(mtd$SampleType)[c(1:8, 10)]) {
    prvl_ls[[i]] <- subset_samples(ps, SampleType == i) %>%
      prevalence(detection = 10, count = TRUE, include.lowest = TRUE)
}

for (i in c("FeedREF", "FeedIM")) {
    prvl_ls[[i]] <- subset_samples(ps, SampleName == i) %>%
      prevalence(detection = 10, count = TRUE, include.lowest = TRUE)
}
```

Compute microbial overlap.

```{r}
ovrl <- bind_cols(prvl_ls) %>%
  pivot_longer("REF-PID":"IM-DIM", 
               names_to = "SampleType", 
               values_to = "Intestine") %>%
  # compute OTU overlap between water/feed and intestinal samples
  mutate(
    WaterOverlap = case_when(
      Water > 0  & Intestine >  0 ~ "Shared",
      Water > 0  & Intestine == 0 ~ "Unique to water",
      Water == 0 & Intestine >  0 ~ "Unique to intestine",
      TRUE ~ "Absent"),
    FeedOverlapREF = case_when(
      FeedREF > 0  & Intestine >  0 ~ "Shared",
      FeedREF > 0  & Intestine == 0 ~ "Unique to feed",
      FeedREF == 0 & Intestine >  0 ~ "Unique to intestine",
      TRUE ~ "Absent"),
    FeedOverlapIM = case_when(
      FeedIM > 0  & Intestine >  0 ~ "Shared",
      FeedIM > 0  & Intestine == 0 ~ "Unique to feed",
      FeedIM == 0 & Intestine >  0 ~ "Unique to intestine",
      TRUE ~ "Absent")) %>%
  pivot_longer(WaterOverlap:FeedOverlapIM, 
               names_to = "OverlapType", 
               values_to = "Category") %>%
  count(OverlapType, SampleType, Category) %>%
  mutate(
    Category = factor(
      Category, 
      levels = c("Unique to water", "Unique to feed", 
                 "Shared", "Unique to intestine", "Absent")),
    Diet = ifelse(grepl("REF", SampleType), "REF", "IM"),
    Diet = factor(Diet, levels = c("REF", "IM")),
    Compartment = ifelse(grepl("PI", SampleType), "PI", "DI"),
    Compartment = factor(Compartment, levels = c("PI", "DI")),
    SampleOrigin = ifelse(grepl("PID|DID", SampleType), "Digesta", "Mucosa"),
    SampleOrigin = factor(SampleOrigin, levels = c("Digesta", "Mucosa")))
```

## Plotting

```{r}
# plot microbial overlap between feed and intestinal samples
ovrl_bar_feed <- ovrl %>%
  filter(OverlapType == "FeedOverlapREF" & Diet == "REF" |
         OverlapType == "FeedOverlapIM"  & Diet == "IM") %>%
  filter(Category != "Absent") %>%
  ggplot(aes(x = SampleType, y = n, fill = Category, label = n)) +
  geom_bar(stat = "identity") +
  geom_text(position = position_stack(vjust = 0.5), size = 3) +
  labs(x = "Sample type", y = "Number of OTUs") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) + 
  scale_fill_manual(values = c("#A6CEE3", "grey", "#1F78B4")) +
  facet_nested(~ Diet + SampleOrigin + Compartment, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "top",
        legend.title = element_blank(),
        legend.margin = margin(0, 0, 0, 0),
        legend.box.margin = margin(5, -10, -10, -10))  

# plot microbial overlap between water and intestinal samples
ovrl_bar_water <- ovrl %>% 
  filter(OverlapType == "WaterOverlap", 
         Category != "Absent") %>%
  ggplot(aes(x = SampleType, y = n, fill = Category, label = n)) +
  geom_bar(stat = "identity") +
  geom_text(position = position_stack(vjust = 0.5), size = 3) +
  labs(x = "Sample type", y = "Number of OTUs") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) + 
  scale_fill_manual(values = c("#B2DF8A", "grey", "#33A02C")) +
  facet_nested(~ Diet + SampleOrigin + Compartment, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "top",
        legend.title = element_blank(),
        legend.margin = margin(0, 0, 0, 0),
        legend.box.margin = margin(5, -10, -10, -10))  
```

Assemble plots.

```{r, fig.width=8}
ovrl_bar <- plot_grid(ovrl_bar_feed, ovrl_bar_water, nrow =  1, align = 'h', rel_widths = c(1, 1)) 
ovrl_bar
```

# Core microbiota

## Compute core microbiota

Data preprocessing.

```{r}
# subset intestinal microbiota samples
ps_int <- ps %>% subset_samples(is.na(Diet) == FALSE) 

# remove features with zero count in all samples
ps_int <- filter_taxa(ps_int, function(x) sum(x > 0) > 0, TRUE) 

# total sum scaling
ps_int_tss <- transform_sample_counts(ps_int, function(x){x / sum(x)}) 
```

Compute core features.

```{r}
core_otu <- core_members(ps_int_tss, detection = 0.001, prevalence = 0.5)
```

Core feature taxonomy.

```{r}
txnm[core_otu, ] %>% 
  select(-Kingdom) %>%
  datatable( 
    options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:6)))
                   )
            )
```

Core feature abundance in each sample.

```{r}
core_otu_abundance <- core(ps_int_tss, detection = 0.001, prevalence = 0.5) %>%
  sample_sums() 

data.frame(core_otu_abundance) %>%
  mutate(core_otu_abundance = round(core_otu_abundance, 2)) %>%
  bind_cols(mtd[sample_names(ps_int_tss), ]) %>%
  select(core_otu_abundance, SampleType) %>%
  datatable( 
  options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:2)))
                 )
            )
```

## Core microbiota heatmap

Make an initial core microbiota heatmap.

```{r, fig.height=8, fig.width=4}
core_htmp <- plot_core(
  ps_int_tss, 
  plot.type = "heatmap", 
  colours = rev(brewer.pal(5, "RdBu")),
  prevalences = seq(0.1, 1, 0.1), 
  detections = 10,
  min.prevalence = 0.5) 

core_htmp <- core_htmp + 
  scale_fill_gradientn(
    colours = rev(brewer.pal(5, "RdBu")),
    breaks  = seq(0, 1, 0.1), 
    labels  = formatC(seq(0, 1, 0.1), format = "f", digits = 1), 
    limits  = c(0, 1)) +
  labs(x = "Detection threshold (%)", y = "OTUs") +
  theme_bw() +
  theme(legend.margin=margin(0,0,0,0), legend.box.margin=margin(0,-5,0,-5))  

core_htmp
```

Format x axis text.

```{r, fig.height=8, fig.width=4}
# get plot data
core_htmp_dt <- core_htmp$data

# format x axis labels
thrs <- unique(core_htmp_dt$DetectionThreshold) %>%
  as.character() %>%
  as.numeric()
  
axis_text_x <- round(thrs * 100, 1) %>% as.character()

# add custom labels
core_htmp <- core_htmp + scale_x_discrete(labels = axis_text_x) 
core_htmp
```

Add taxonomy to the y axis text.

```{r, fig.height=8, fig.width=6}
# add the taxonomy of core OTUs
core_htmp_dt1 <- txnm[levels(core_htmp_dt$Taxa), ] %>%
  rownames_to_column("Taxa") %>%
  inner_join(core_htmp_dt, by = "Taxa") %>%
  # get the best taxonomic annotations for core OTUs
  mutate(
    Class = ifelse(is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
                   Phylum, 
                   Class),
    Order = ifelse(is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
                   Class, 
                   Order),
    Family = ifelse(is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
                    Order, 
                    Family),
    Genus = ifelse(is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
                   Family, 
                   Genus)) %>%
  select(Taxa, Genus, DetectionThreshold, Prevalence) %>%
  # italicize taxa names using R expressions
  mutate(Taxa = paste0(Taxa, ":italic(", Genus, ")"),
         # some taxa names contain space; tilde (~) is recognized as "space" in R expressions
         Taxa = gsub("\\s+", "~", Taxa), 
         Taxa = gsub("g__", "", Taxa),
         Taxa = factor(Taxa, levels = unique(Taxa))) %>%
  select(-Genus)

# replace the data
core_htmp$data <- core_htmp_dt1

# parse y axis text as R expression
axis_text_y <- parse(text = gsub("OTU.*:", "", levels(core_htmp_dt1$Taxa))) 

# make a new plot
core_htmp <- core_htmp + scale_y_discrete(labels = axis_text_y)
core_htmp
```

# Figure 1

Assemble plots.

```{r, fig.width=14, fig.height=8}
fig1_ab <- plot_grid(taxa_bar, ovrl_bar, ncol =  1, align = 'v',  
                     rel_heights = c(1.2, 1), labels = c("a", "b")) 

fig1 <- plot_grid(fig1_ab, core_htmp, ncol =  2, align = 'h', 
                  rel_widths = c(1.6, 1), labels = c("", "c"))

fig1
```

Export plot

```{r}
ggsave(here("result/figure/Figure 1.tiff"), fig1, width = 14, height = 8,
       units = "in", dpi = 300, compression = "lzw")
```

# Session information

```{r session-info}
sessionInfo()
```
