---
title: "Taxonomic analysis"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = FALSE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "left",  buttonSide = "left")`

# Getting ready

Load packages.

```{r load-package}
library(here) # A Simpler Way to Find Your Files, CRAN v1.0.1
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN v1.3.0 
library(cowplot) # Streamlined Plot Theme and Plot Annotations for 'ggplot2', CRAN v1.1.1  
library(ggpubr) # 'ggplot2' Based Publication Ready Plots, CRAN v0.4.0
library(ggh4x) # Hacks for 'ggplot2', CRAN v0.1.2.1 
library(ggtext) # Improved Text Rendering Support for 'ggplot2', CRAN v0.1.1
library(PerformanceAnalytics) # Econometric Tools for Performance and Risk Analysis, CRAN v2.0.4
library(RColorBrewer) # ColorBrewer Palettes, CRAN v1.1-2 
library(qiime2R) # Import qiime2 artifacts to R, [github::jbisanz/qiime2R] v0.99.35  
library(MicrobeR) # Handy functions for microbiome analysis in R, [github::jbisanz/MicrobeR] v0.3.2 
library(microbiome) # Microbiome Analytics, Bioconductor v1.12.0
library(speedyseq) # Faster implementations of phyloseq functions, [github::mikemc/speedyseq] v0.5.3.9001  #
library(flextable) # Functions for Tabular Reporting, CRAN v0.6.1  
library(officer) # Manipulation of Microsoft Word and PowerPoint Documents, CRAN v0.3.16     
library(DT) # A Wrapper of the JavaScript Library 'DataTables', CRAN v0.17 
```

Load data.

```{r}
# load phyloseq object
ps <- readRDS(here("data/intermediate/qiime2R/phyloseq.rds"))

# filter samples
ps <- ps %>%
  # remove control samples
  subset_samples(!SampleType %in% c("Extraction-blank", "PCR-blank", "Mock")) %>%
  # remove technical replicates
  subset_samples(!(IsTechnicalReplicate == "yes" & PCRBatch == 2))


# change OTU names for easy display
indx <- formatC(1:ntaxa(ps), width = nchar(ntaxa(ps)), format = "d", flag = "0")
taxa_names(ps) <- paste0("OTU", indx)

# extract otu table, taxonomy and metadata
otu <- as.data.frame(otu_table(ps)) 
txnm <- tax_table(ps) %>% as("matrix") %>% as.data.frame()
mtd <- data.frame(sample_data(ps), check.names = FALSE)
```

# Taxonomic assignment overview

First of all, let's look at the coverage of taxonomic assignments at different rank levels.

```{r}
txnm_assg <- txnm %>%
  gather("Rank", "Name", rank_names(ps)) %>%
  group_by(Rank) %>%
  # Empty taxonomic ranks may be na or strings containing "uncultured" or "Ambiguous_taxa"
  summarize(OTUs_classified = sum(!is.na(Name) & !grepl("uncultured|Ambiguous|metagenome", Name))) %>%
  mutate(
    Frac_classified = OTUs_classified / ntaxa(ps),
    Frac_classified = ifelse(
      Frac_classified == 1, 
      "100", 
      round(Frac_classified * 100, 1)),
    Frac_classified = paste(Frac_classified, "%"),
    Rank = factor(Rank, rank_names(ps))) %>%
  arrange(Rank) 

txnm_assg %>%
  datatable(options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:3)))))
```

We can tell that the majority of OTUs were assigned at the genus level whereas only `r txnm_assg[7, 3]` of OTUs got a species-level annotation.

# Taxa barplot

## Mock

### Expected mock composition

```{r}
mock_exp <- read_tsv(here("data/reference/mock_expected.tsv")) 

mock_exp <- rename(mock_exp[, 1:2], taxa_exp = 1, abnd_exp = 2) %>%
  arrange(taxa_exp) %>%
  mutate(taxa_exp = gsub("D_0.*D_6__", "", taxa_exp)) # prune taxonomy to the lowest level
```

### Observed mock composition

```{r}
mock_obs <- read_qza(here("data/intermediate/qiime2/97otu/quality-control/mock-observed-l7-rel.qza"))
mock_obs <- mock_obs$data %>%
  as.data.frame() %>%
  rownames_to_column("tax_obs") %>%
  # prune taxonomy to the lowest level
  mutate(tax_obs = gsub("D_0.*D_6__", "", tax_obs), # species level  
         tax_obs = gsub("D_0.*D_5__|;__", "", tax_obs)) # genus level
```

### Merge data

```{r}
mock <- bind_cols(mock_exp, mock_obs) %>%
  mutate(
    gram_staining = c("(G+)", "(G+)", "(G+)", "(G+)", "(G+)", "(G-)", "(G-)", "(G-)"),
    taxa = paste0("*", taxa_exp, "*", " ", gram_staining, " / ", "*", tax_obs, "*")) %>% 
  select(-taxa_exp, -gram_staining, -tax_obs) %>%
  mutate(taxa = factor(taxa, levels = unique(taxa))) %>% 
  # use percentage as unit for relative abundance
  mutate_if(is.numeric, ~(100 * .x)) %>% 
  mutate_if(is.numeric, ~round(.x, 1))
```

### Correlation plot

```{r}
mock[, !(colnames(mock) == "taxa")] %>%
  rename(Expected = abnd_exp) %>%
  chart.Correlation(histogram = FALSE, pch = 19)
```

### Stacked barplot

```{r, fig.width=8, fig.height=5}
bar_mock <- mock %>% 
  gather(key = "type", value = "abundance", -taxa) %>%
  ggplot(aes(x = type, y = abundance, fill = taxa, label = abundance)) +
  geom_bar(stat = "identity") +
  geom_text(position = position_stack(vjust = 0.5), size = 4) +
  annotate("segment", x = 1.75, xend = 6.25, y = 102, yend = 102) +
  annotate(
    "text", x = c(1, 4.5), y = c(105, 105), 
    label = c("Expected", "Observed"), size = 4
    ) +
  scale_y_continuous(
    limits = c(0, 105), 
    breaks = 0:10*10, 
    expand = expansion(mult = c(0, 0.05))
    ) +
  labs(
    x = "Mock samples", 
    y = "Relative abundance (%)", 
    fill = "Taxa (expected (Gram stain) / observed)"
    ) +
  scale_fill_manual(
    values = brewer.pal(8, "Paired"), 
    labels = levels(mock$taxa)
    ) +
  theme_minimal() +
  theme(
    axis.text.x = element_blank(),
    legend.text = element_markdown()) # align legend text to left

bar_mock
```

### Figure S1

```{r}
ggsave(here("result/figure/Figure S1.tiff"), width = 8, height = 5, 
       units = "in", dpi = 300, compression = "lzw")
```

## Biological samples

### Top10 genera

Get top 10 most abundant taxa at genus level.

```{r}
taxa_tab <- Summarize.Taxa(otu, txnm)$Genus %>% Make.Percent() 
taxa_tab <- taxa_tab[order(rowMeans(taxa_tab), decreasing = T), ]
Others <- colSums(taxa_tab[11:nrow(taxa_tab), ])
taxa_tab <- rbind(taxa_tab[1:10, ], Others)
```

Tidy taxa table for making stacked bar plots.

```{r}
taxa_tab <- as.data.frame(taxa_tab) %>%
  rownames_to_column("Taxa") %>%
  separate(
    Taxa, 
    sep = ";", 
    c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")) %>% 
  mutate(
    Phylum = ifelse(
      is.na(Phylum)|Phylum == "NA"|grepl("uncultured|Ambiguous|metagenome", Phylum), 
      Kingdom, 
      Phylum),
    Class = ifelse(
      is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
      Phylum, 
      Class),
    Order = ifelse(
      is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
      Class, 
      Order),
    Family = ifelse(
      is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
      Order, 
      Family),
    Genus = ifelse(
      is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
      Family, 
      Genus)) %>%
  select(-Kingdom, -(Class:Family)) %>%
  mutate(
    Phylum = gsub("p__", "", Phylum),
    Phylum = factor(Phylum, levels = rev(unique(Phylum))),
    Genus = gsub("g__", "", Genus),
    Genus = factor(Genus, levels = rev(unique(Genus)))) %>%
  arrange(Phylum, Genus) %>%
  mutate(Genus = factor(Genus, levels = unique(Genus))) %>%
  pivot_longer(-c(Phylum, Genus), names_to = "SampleID", values_to = "Abundance") %>%
  inner_join(rownames_to_column(mtd, "SampleID"), by = "SampleID")
```

### Stacked barplot

```{r, fig.width=10}
# define color scheme
col <- c("grey", brewer.pal(n = 10, name = "Paired"))

# digesta samples
taxa_bar_digesta <- filter(taxa_tab, SampleOrigin == "Digesta") %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "Relative abundance (%)") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_nested(
    ~ SampleOrigin + Diet + Compartment, 
    scales = "free_x", 
    nest_line = TRUE
    ) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# mucosa samples  
taxa_bar_mucosa <- filter(taxa_tab, SampleOrigin == "Mucosa") %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_nested(
    ~ SampleOrigin + Diet + Compartment, 
    scales = "free_x", 
    nest_line = TRUE
    ) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# feed samples
taxa_bar_feed <- filter(taxa_tab, SampleType == "Feed") %>%
  mutate(
    Diet = ifelse(SampleName == "FeedREF", "REF", "IM"),
    Diet = factor(Diet, levels = c("REF", "IM"))
    ) %>%
  ggplot(aes(x = SampleName, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity", width = 0.5) +
  labs(x = "", y = "") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_nested(~ SampleOrigin + Diet, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# water samples
taxa_bar_water <- filter(taxa_tab, SampleType == "Water") %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "") +
  scale_y_continuous(breaks = 0:10*10, expand = c(0,0)) + 
  scale_fill_manual(values = col) +
  facet_wrap(~ SampleOrigin) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") 

# assemble plots
taxa_bar <- plot_grid(
  taxa_bar_digesta + theme(plot.margin = margin(b = -0.5, unit = "cm")), 
  taxa_bar_mucosa  + theme(plot.margin = margin(l = 0, b = -0.5, unit = "cm")),
  taxa_bar_feed    + theme(plot.margin = margin(l = -0.5, b = -0.5, unit = "cm")),
  taxa_bar_water   + theme(plot.margin = margin(l = -0.5, b = -0.5, unit = "cm")), 
  nrow =  1, align = 'h', axis = "tb", rel_widths = c(7.5, 7.5, 1.6, 1.4)) 

# make a joint x axis label for the assembled plot
xlab <- ggdraw() + 
  draw_label("Samples", fontface = "plain") + 
  theme(plot.margin = margin(t = -0.5, unit = "cm"))

# add the x axis label
taxa_bar <- plot_grid(taxa_bar, xlab, ncol = 1, rel_heights = c(10, 1))
taxa_bar
```

### Modify legend

Modify legend using flextable.

```{r}
lgnd <- unique(taxa_tab[ , c("Genus", "Phylum")]) %>%
  #arrange(factor(Genus, levels = levels(taxa_tab$Genus))) %>%
  mutate(col = 10, .before = Genus) %>% # legend color
  flextable() %>% 
  theme_box() %>%
  highlight(j = "col", color = col, part = "body") %>% # use legend color to highlight text
  color(j = "col", color = col, part = "body") %>% # use legend color as font color to hide the text
  color(j = "col", color = "white", part = "header") %>% # head text in the header
  italic(i = 2:11, italic = TRUE, part = "body") %>% # italicize taxa names
  fontsize(size = 20, part = "all") %>% # font size
  width(j = c("col"), width = 0.2) %>% # change the width of legend
  height_all(height = 1) %>% # cell height
  align(j = c("Genus", "Phylum"), align = "left", part = "all") %>% # align text to the left
  merge_v(j = "Phylum") %>% # merge cells belong to the same Phylum
  border_remove() %>% # remove borders
  theme_booktabs() %>% # apply table style
  hline(i = c(1, 2), border = fp_border(width = 2), part = "body") %>% # draw horizontal lines
  hline_top(part = "header", border = fp_border(color = "white")) # remove horizontal line in the table header

# convert flextable into a ggplot object
lgnd <- as_raster(lgnd, webshot = "webshot")

lgnd <- ggplot() + 
  theme_void() + 
  annotation_custom(grid::rasterGrob(lgnd, y = unit(0.5, "npc"))) # top right: y = unit(0.72, "npc")
```

### Figure 1

```{r, fig.width=10, fig.height=6}
taxa_bar <- plot_grid(taxa_bar, lgnd, nrow =  1, align = 'h', rel_widths = c(4, 1)) 
taxa_bar

# export plot
ggsave(here("result/figure/Figure 1.tiff"), width = 10, height = 6,
       units = "in", dpi = 300, compression = "lzw")
```

# Microbial overlap

## Number of OTUs shared

### Compute OTU prevalence

```{r}
# total sum scaling
ps_tss <- transform_sample_counts(ps, function(x){x / sum(x)})

# compute OTU prevalence
prvl <- list()

for (i in c("REF-PIM", "REF-DIM", "IM-PIM", "IM-DIM", "Water")) {
    prvl[[i]] <- subset_samples(ps_tss, SampleType == i) %>%
      prevalence(detection = 0.001, count = TRUE, include.lowest = TRUE)
}

for (i in c("FeedREF", "FeedIM")) {
    prvl[[i]] <- subset_samples(ps_tss, SampleName == i) %>%
      prevalence(detection = 0.001, count = TRUE, include.lowest = TRUE)
}
```

### Compute OTU overlap

```{r}
ovrl <- bind_cols(prvl) %>%
  pivot_longer("REF-PIM":"IM-DIM", 
               names_to = "SampleType", 
               values_to = "Intestine") %>%
  # compute OTU overlap between water/feed and intestinal samples
  mutate(
    WaterOverlap = case_when(
      Water >  0 & Intestine >  0 ~ "Shared",
      Water >  0 & Intestine == 0 ~ "Unique to water",
      Water == 0 & Intestine >  0 ~ "Unique to intestine",
      TRUE ~ "Absent"),
    FeedOverlapREF = case_when(
      FeedREF == 1 & Intestine >  0 ~ "Shared",
      FeedREF == 1 & Intestine == 0 ~ "Unique to feed",
      FeedREF == 0 & Intestine >  0 ~ "Unique to intestine",
      TRUE ~ "Absent"),
    FeedOverlapIM = case_when(
      FeedIM == 1 & Intestine >  0 ~ "Shared",
      FeedIM == 1 & Intestine == 0 ~ "Unique to feed",
      FeedIM == 0 & Intestine >  0 ~ "Unique to intestine",
      TRUE ~ "Absent")) %>%
  pivot_longer(WaterOverlap:FeedOverlapIM, 
               names_to = "OverlapType", 
               values_to = "Category") %>%
  count(OverlapType, SampleType, Category) %>%
  mutate(
    Category = factor(
      Category, 
      levels = c("Unique to water", "Unique to feed", 
                 "Shared", "Unique to intestine", "Absent")),
    Diet = ifelse(grepl("REF", SampleType), "REF", "IM"),
    Diet = factor(Diet, levels = c("REF", "IM")),
    Site = ifelse(grepl("PIM", SampleType), "PIM", "DIM"),
    Site = factor(Site, levels = c("PIM", "DIM")))
```

### Plotting

Plot microbial overlap between feed and intestinal samples.

```{r}
ovrl_bar_feed <- ovrl %>%
  filter(OverlapType == "FeedOverlapREF" & Diet == "REF" |
         OverlapType == "FeedOverlapIM"  & Diet == "IM") %>%
  filter(Category != "Absent") %>%
  ggplot(aes(x = SampleType, y = n, fill = Category, label = n)) +
  geom_bar(stat = "identity") +
  geom_text(position = position_stack(vjust = 0.5), color = "white", size = 4) +
  labs(x = "Comparment", y = "Number of OTUs") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) + 
  scale_fill_manual(values = c("#B2DF8A", "grey50", "#33A02C")) +
  facet_nested(~ Diet + Site, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.title = element_blank(),
        legend.margin = margin(0, 0, 0, 0),
        legend.box.margin = margin(0, 0, 0, -5))  
```

Plot microbial overlap between water and intestinal samples.

```{r}
ovrl_bar_water <- ovrl %>% 
  filter(OverlapType == "WaterOverlap", 
         Category != "Absent") %>%
  ggplot(aes(x = SampleType, y = n, fill = Category, label = n)) +
  geom_bar(stat = "identity") +
  geom_text(position = position_stack(vjust = 0.5), color = "white", size = 4) +
  labs(x = "Comparment", y = "Number of OTUs") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) + 
  scale_fill_manual(values = c("#A6CEE3", "grey50", "#1F78B4")) +
  facet_nested(~ Diet + Site, scales = "free_x", nest_line = TRUE) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.title = element_blank(),
        legend.margin = margin(0, 0, 0, 0),
        legend.box.margin = margin(0, 0, 0, -5))  
```

Assemble plots.

```{r, fig.width=8, fig.height=4}
plot_grid(ovrl_bar_feed, ovrl_bar_water, nrow =  1) 
```

## Abundance of OTUs shared

### Compute shared OTU abundance 

Compute the abundance of shared OTUs between feed and intestinal mucosa.

```{r}
shared_feed <- map(c("FeedREF", "FeedIM"), function(x) {
  prune_taxa(names(which(prvl[[x]] == 1)), ps_tss) %>% sample_sums()
  }
)

names(shared_feed) <- c("REF", "IM")
  
# combine with metadata
shared_feed <- map(seq_along(shared_feed), function(x){
  data.frame(Abundance = shared_feed[[x]]) %>%
  rownames_to_column("SampleID") %>%
  inner_join(rownames_to_column(mtd, "SampleID"), by = "SampleID") %>%
  filter(Diet == names(shared_feed)[x] & !SampleOrigin %in% c("Feed", "Water")) %>%
  mutate(Compartment = ifelse(Compartment == "PI", "PIM", "DIM"),
         Compartment = factor(Compartment, levels = c("PIM", "DIM"))) %>%
  mutate(Abundance = round(100 * Abundance, 1))}) %>%
  bind_rows()
```

Compute the abundance of shared OTUs between water and intestinal mucosa.

```{r}
shared_water <- prune_taxa(names(which(prvl[["Water"]] > 0)), ps_tss) %>% 
  sample_sums()

# combine with metadata
shared_water <- data.frame(Abundance = shared_water) %>%
  rownames_to_column("SampleID") %>%
  inner_join(rownames_to_column(mtd, "SampleID"), by = "SampleID") %>%
  filter(!SampleOrigin %in% c("Feed", "Water")) %>%
  mutate(Compartment = ifelse(Compartment == "PI", "PIM", "DIM"),
         Compartment = factor(Compartment, levels = c("PIM", "DIM"))) %>%
  mutate(Abundance = round(100 * Abundance, 1))
```

### Plotting

Abundance of shared OTUs between feed and intestinal mucosa.

```{r}
bp_shared_feed <- ggplot(shared_feed, aes(x = Compartment, y = Abundance)) +
  geom_boxplot(fill = "grey50", width = 0.5) +
  geom_point() +
  facet_nested(~ Diet + Compartment, scales = "free_x", nest_line = TRUE) +
  labs(x = "Comparment", y = "Relative abundance of shared OTUs (%)") +
  scale_y_continuous(
    limits = c(0, 100), breaks = 0:10*10, expand = expansion(c(0, 0))
    ) + 
  theme_bw() +
  theme(axis.text.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "none")
```

Abundance of shared OTUs between water and intestinal mucosa

```{r}
bp_shared_water <- shared_water %>%
  ggplot(aes(x = Compartment, y = Abundance)) +
    geom_boxplot(fill = "grey50", width = 0.5) +
    geom_point() +
    facet_nested(~ Diet + Compartment, scales = "free_x", nest_line = TRUE) +
    labs(x = "Comparment", y = "Relative abundance of shared OTUs (%)") +
    scale_y_continuous(
      limits = c(0, 100), breaks = 0:10*10, expand = expansion(c(0, 0))
      ) + 
    theme_bw() +
    theme(axis.text.x = element_blank(),
          strip.background = element_blank(),
          legend.position = "none") 
```

Assemble plots.

```{r, fig.width=6, fig.height=4}
plot_grid(bp_shared_feed, bp_shared_water, nrow =  1) 
```

## Figure 2

```{r, fig.width=7, fig.height=8}
plot_grid(ovrl_bar_feed, bp_shared_feed, ovrl_bar_water, bp_shared_water, 
          ncol = 2, align = "h", axis = "tb", rel_widths = c(1.5, 1), 
          labels = c("a", "", "b", "")) 

# export plot
ggsave(here("result/figure/Figure 2.tiff"), width = 7, height = 8,
       units = "in", dpi = 300, compression = "lzw")
```

# Core microbiota

## Compute core microbiota

Data preprocessing.

```{r}
# subset phyloseq
ps_core <- list(
  REF = subset_samples(ps, Diet == "REF" & !SampleOrigin %in% c("Feed", "Water")),
  IM = subset_samples(ps, Diet == "IM" & !SampleOrigin %in% c("Feed", "Water"))
  )

# data filtering and transformation
ps_core <- ps_core %>%
  # remove features with zero count in all samples
  map(~filter_taxa(.x, function(x) sum(x > 0) > 0, TRUE)) %>%
  # aggregate OTUs at genus level
  #map(~tax_glom(.x, "Genus")) %>%
  # total sum scaling
  map(~transform_sample_counts(.x, function(x){x / sum(x)}))
```

## Core microbiota heatmap

### Initial heatmap

```{r, results='hide'}
# basic plot
coreHmp <- map(seq_along(ps_core), function(x){
  if (names(ps_core)[x] == "REF"){
    plot_core(ps_core[[x]], 
      plot.type = "heatmap", 
      colours = rev(brewer.pal(5, "RdBu")),
      prevalences = seq(0.1, 1, 0.1), 
      detections = 9,
      min.prevalence = 0.5) 
  } else {
    plot_core(ps_core[[x]], 
      plot.type = "heatmap", 
      colours = rev(brewer.pal(5, "RdBu")),
      prevalences = seq(0.1, 1, 0.1), 
      detections = 8,
      min.prevalence = 0.5) 
    }
  }
)

names(coreHmp) <- c("REF", "IM")  

# change aesthetics
coreHmp <- map(seq_along(coreHmp), function(x){
  coreHmp[[x]] + 
  scale_fill_gradientn(
    colours = rev(brewer.pal(5, "RdBu")),
    breaks = seq(0, 1, 0.1), 
    labels = formatC(seq(0, 1, 0.1), format = "f", digits = 1), 
    limits = c(0, 1)) +
  labs(title = names(coreHmp)[x], x = "Detection threshold (%)", y = "") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))
  }
)  
```

### Format x-axis text

```{r}
# get plot data
coreHmp_dt <- map(coreHmp, ~.x$data)

# format x axis labels
xlab <- map(coreHmp_dt, function(x){
  unique(x$DetectionThreshold) %>% 
    as.character() %>% 
    as.numeric()
  }
) 

xlab <- map(xlab, ~formatC(.x * 100, format = "f", digits = 1))   

# add custom labels
coreHmp <- map2(coreHmp, xlab, function(x, y) x + scale_x_discrete(labels = y)) 
```

### Format y-axis text

```{r}
# add the taxonomy of core taxa
coreHmp_dt <- map(coreHmp_dt, function(x){
  txnm[levels(x$Taxa), ] %>%
  rownames_to_column("Taxa") %>%
  inner_join(x, by = "Taxa") %>%
  # get the best taxonomic annotations for core taxa
  mutate(
    Class = ifelse(
      is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
      Phylum, 
      Class),
    Order = ifelse(
      is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
      Class, 
      Order),
    Family = ifelse(
      is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
      Order, 
      Family),
    Genus = ifelse(
      is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
      Family, 
      Genus)) %>%
  select(Taxa, Genus, DetectionThreshold, Prevalence) %>%
  # italicize taxa names using R expressions
  mutate(Taxa = paste0(Taxa, ":italic(", Genus, ")"),
         # some taxa names contain space; tilde (~) is recognized as "space" in R expressions
         Taxa = gsub("\\s+", "~", Taxa), 
         Taxa = gsub("g__", "", Taxa),
         Taxa = factor(Taxa, levels = unique(Taxa))) %>%
  select(-Genus)
  }
)

# replace the data
coreHm <- map2(coreHmp, coreHmp_dt, function(x, y) x$data <- y)

# parse y axis text as R expression
ylab <- map(coreHmp_dt, ~parse(text = gsub("OTU.*:", "", levels(.x$Taxa))))  

# make new plots
coreHmp <- map2(coreHmp, ylab, function(x, y) x + scale_y_discrete(labels = y)) 
```

### Figure 3

```{r, fig.width=8.5, fig.height=10}
coreHmp <- ggarrange(
  plotlist = coreHmp, 
  nrow = 1, 
  labels = c("OTUs", ""), 
  label.y = 0.55, 
  font.label = list(size = 10, face = "plain"), 
  legend = "right", 
  common.legend = TRUE)

coreHmp

# export plot
ggsave(here("result/figure/Figure 3.tiff"), width = 8.5, height = 10,
       units = "in", dpi = 300, compression = "lzw")
```


# Session information

```{r session-info}
sessionInfo()
```
