---
title: "Multivariable association analysis"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  pdf_document: 
    latex_engine: xelatex
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
options(width = 120)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = TRUE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "left",  buttonSide = "left")`

# Getting ready

Load packages.

```{r}
library(here) # A Simpler Way to Find Your Files, CRAN v1.0.1
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN v1.3.0 
library(RColorBrewer) # ColorBrewer Palettes, CRAN v1.1-2 
library(phyloseq) # Handling and analysis of high-throughput microbiome census data, Bioconductor v1.34.0 
library(MicrobeR) # Handy functions for microbiome analysis in R, [github::jbisanz/MicrobeR] v0.3.2 
library(Maaslin2) # Maaslin2, Bioconductor v1.4.0
library(patchwork) # The Composer of Plots, CRAN v1.1.1  
```

Load function.

```{r}
source(here("code/functions/plot_heatmap.R"))
```

Set ggplot global theme.

```{r}
theme_set(
  theme_bw() +
  theme(legend.position = "top", strip.text = element_text(face = "italic")))
```

Import data.

```{r}
ps <- readRDS(here("data/intermediate/qiime2R/phyloseq.rds"))
```

# Data preprocessing

## Metadata, OTU table and taxonomy

```{r}
# filter samples
ps <- ps %>%
  subset_samples(SampleOrigin %in% c("Feed", "Digesta", "Mucosa")) %>%
  subset_samples(!(IsTechnicalReplicate == "yes" & PCRBatch == 2))

# extract metadata, OTU table and taxonomy 
mtd <- data.frame(sample_data(ps), check.names = FALSE) %>%
  mutate(FishID = as.character(FishID)) 

otu <- as.data.frame(otu_table(ps)) 

txnm <- tax_table(ps) %>% as("matrix") %>% as.data.frame()
```

## Collapse OTU table at genus-level

```{r}
taxa_l6 <- Summarize.Taxa(otu, txnm)$Genus %>%
  rownames_to_column("Taxa") %>%
  separate(
    Taxa, 
    sep = ";", 
    c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")) %>% 
  mutate(
    Class = ifelse(
      is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
      Phylum, 
      Class),
    Order = ifelse(
      is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
      Class, 
      Order),
    Family = ifelse(
      is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
      Order, 
      Family),
    Genus = ifelse(
      is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
      Family,
      Genus)) %>%
  select(-(Kingdom:Family)) %>%
  # remove taxa rank prefix "g__" and square brackets in the taxa names
  mutate(Genus = gsub("g__|\\[|\\]", "", Genus)) %>%
  # merge rows with the same taxa name
  group_by(Genus) %>%
  summarise(across(everything(), sum)) %>%
  column_to_rownames("Genus")
```

# Differential abundance testing

Define fixed and random effects.

```{r}
# fixed effects
fixef <- c("Diet", "Compartment") 

# random effects
ranef <- c("FishID", "Tank")
```

## Digesta

### Subset data

```{r}
# metadata
mtd_dgs <- filter(mtd, SampleOrigin == "Digesta")  

# feature table
taxa_l6_dgs <- select(taxa_l6, rownames(mtd_dgs))
```

### Run MaAsLin2.

```{r, results='hide'}
fit_dgs <- Maaslin2(
  input_data = taxa_l6_dgs, 
  input_metadata = mtd_dgs,
  output = here("data/intermediate/maaslin2/digesta"),
  min_abundance = 0,
  min_prevalence = 0.25,
  max_significance = 0.05,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = fixef,
  random_effects = ranef,
  correction = "BH",
  standardize = FALSE,  
  plot_heatmap = TRUE,
  plot_scatter = FALSE,
  cores = parallel::detectCores()
  )
```

### Modify MaAsLin2 outputs

Merge maaslin2 outputs, feature table and metadata.

```{r}
# extract maaslin2 output
fit_dgs_results <- filter(fit_dgs$results) %>%
  mutate(
    metadata = factor(metadata, levels = fixef),
    # fix taxa names in the maaslin2 outputs
    # space (" ") in the original taxa names was replaced with dot(".") in the maaslin2 outputs
    feature = gsub("\\.", " ", feature))

# add feed samples to the feature table
mtd_dgs_fd <- filter(mtd, SampleOrigin %in% c("Feed", "Digesta")) 
taxa_l6_dgs_fd <- select(taxa_l6, rownames(mtd_dgs_fd)) %>%
  mutate(across(everything(), ~.x/sum(.x)))

# filter features showing significant associations with metadata
sigAss_dgs <- filter(fit_dgs_results, qval <= 0.05) %>%
  mutate(N.not.zero = paste0(N.not.zero, "/", N)) %>%
  # NB: taxa names in the maaslin2 outputs may not match those in the feature table!!
  inner_join(rownames_to_column(taxa_l6_dgs_fd, "feature"), by = "feature") %>%
  pivot_longer(
    cols = contains("AqFl1-"), # SampleID starts with "AqFl1-" 
    names_to = "SampleID", 
    values_to = "Abundance") %>%
  inner_join(rownames_to_column(mtd_dgs_fd, "SampleID"), by = "SampleID") %>%
  mutate(
    qval = ifelse(qval < 0.001, "< 0.001", round(qval, 3)), 
    # text annotation 
    ann_text = paste("q-value:", qval)
  ) 
```

### Heatmap

```{r, fig.width=4, fig.height=8}
hmp_dgs <- plot_heatmap(
  maaslin2_output = fit_dgs_results,
  first_n = 50, # plot all data: n_distinct(sigAss_dgs$feature)
  cell_value = "qval",
  data_label = "data",
  metadata_label = "metadata",
  plot_title = FALSE,
  legend_title = FALSE,
  legend_title_position = "topcenter", # topleft, topcenter, leftcenter-rot, lefttop-rot
  legend_direction = "horizontal", # "vertical", "horizontal"
  color = c("blue", "grey90", "red"),
  board_line_col = "white",
  colnames_rotate = 45,
  colnames_fontsize = 12,
  rownames_fontsize = 10,
  italize_rownames = TRUE)

draw(
  hmp_dgs, 
  padding = unit(c(2, 2, 2, 20), "mm"), # bottom, left, top and right
  heatmap_legend_side = "top"
 )
```

### Scatter plot

#### Diet

```{r, fig.width=11, fig.height=38}
# filter features showing significant associations with Diet
sigAss_dgs_diet <- sigAss_dgs %>% 
  filter(metadata == "Diet" & SampleOrigin != "Feed") %>%
  # add the number of non-zero observations in each diet to the text annotation
  group_by(feature, Diet) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(text_extra = paste0("N.not.zero(", Diet, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(text_extra = paste0(text_extra, collapse = "\n")) %>%
  inner_join(filter(sigAss_dgs, metadata == "Diet"), by = "feature") %>%
  mutate(ann_text = paste0(ann_text, "\n", text_extra))

# plotting
filter(sigAss_dgs_diet, SampleOrigin == "Digesta") %>%
ggplot(aes(x = Diet, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Compartment), position = position_jitter(0.2)) +
  # highlight taxa abundance in the feed samples
  geom_point(
    data = filter(sigAss_dgs_diet, SampleOrigin == "Feed"),
    color = "#666666", size = 3) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_text), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

#### Compartment

```{r, fig.width=11, fig.height=12}
# filter features showing significant associations with Compartment
sigAss_dgs_cmprt <- sigAss_dgs %>% 
  filter(metadata == "Compartment" & SampleOrigin != "Feed")

# add the number of non-zero observations to the text annotation
sigAss_dgs_cmprt <- sigAss_dgs_cmprt %>%
  group_by(feature, Compartment) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(text_extra = paste0("N.not.zero(", Compartment, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(text_extra = paste(text_extra, collapse = "\n")) %>%
  inner_join(sigAss_dgs_cmprt, by = "feature") %>%
  mutate(ann_text = paste0(ann_text, "\n", text_extra))

# plotting
ggplot(sigAss_dgs_cmprt, aes(x = Compartment, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Diet), position = position_jitter(0.2)) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_text), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

### Figure 6

### Figure 6b

Highlight representative taxa enriched in fish fed the REF diet.

```{r, fig.width=6.5, fig.height=5}
# taxa to be shown in the subplot
fig6b_taxa_name <- c(
  "Lactobacillus", "f__Peptostreptococcaceae", "Peptostreptococcus", 
  "Lactococcus", "Leuconostoc", "Pediococcus"
  )

# plot
fig6b <- sigAss_dgs_diet %>% 
  filter(SampleOrigin == "Digesta" & feature %in% fig6b_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig6b_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Compartment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_dgs_diet %>%
        filter(SampleOrigin == "Feed" & feature %in% fig6b_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig6b_taxa_name)), 
      color = "#666666", size = 3) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_text), size = 3, hjust = 1, vjust = 1) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 1)) +   
    labs(y = "Relative abundance") +
    scale_color_brewer(palette = "Dark2") 

fig6b 
```

### Figure 6c

Highlight representative taxa enriched in fish fed the IM diet.

```{r, fig.width=6.5, fig.height=5}
# taxa to be shown in the subplot
fig6c_taxa_name <- c(
  "Oceanobacillus", "Bacillus", "f__Bacillaceae",  
  "Ornithinibacillus", "Corynebacterium 1", "Enterococcus"
  )

# plot
fig6c <- filter(
  sigAss_dgs_diet, 
  SampleOrigin == "Digesta" & feature %in% fig6c_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig6c_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Compartment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_dgs_diet %>%
        filter(SampleOrigin == "Feed" & feature %in% fig6c_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig6c_taxa_name)), 
      color = "#666666", size = 3) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_text), size = 3, hjust = 1, vjust = 1) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 1.0)) +   
    labs(y = "Relative abundance") +
    scale_color_brewer(palette = "Dark2") 

fig6c 
```

### Figure 6

Assemble plots.

```{r, fig.width=10, fig.height=12}
# convert the heatmap into a "grob"
fig6_main <- grid.grabExpr(
  draw(
    hmp_dgs, 
    padding = unit(c(2, 2, 2, 8), "mm"), # bottom, left, top and right
    heatmap_legend_side = "top"
   )
  ) 

# assemble subplots
fig6_sub <- fig6b / fig6c + theme(plot.margin = margin(b = 1.5, unit = "cm")) +
  plot_layout(guides = 'collect') 

# heatmap + subplots
wrap_elements(full = fig6_main) + fig6_sub +
  plot_annotation(tag_levels = "a") +
  plot_layout(widths = c(1, 1.6)) 

# export plot
ggsave(here("result", "figure", "Figure 6.tiff"), width = 10, height = 12,
       units = "in", dpi = 300, compression = "lzw")
```

## Mucosa

### Subset data

```{r}
# metadata
mtd_mcs <- filter(mtd, SampleOrigin == "Mucosa")  

# feature table
taxa_l6_mcs <- select(taxa_l6, rownames(mtd_mcs))
```

### Run MaAsLin2.

```{r, results='hide'}
fit_mcs <- Maaslin2(
  input_data = taxa_l6_mcs, 
  input_metadata = mtd_mcs,
  output = here("data/intermediate/maaslin2/mucosa"),
  min_abundance = 0,
  min_prevalence = 0.25,
  max_significance = 0.05,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = fixef,
  random_effects = ranef,
  correction = "BH",
  standardize = FALSE,  
  plot_heatmap = TRUE,
  plot_scatter = FALSE,
  cores = parallel::detectCores()
  )
```

### Modify MaAsLin2 outputs

Merge maaslin2 outputs, feature table and metadata.

```{r}
# extract maaslin2 output
fit_mcs_results <- filter(fit_mcs$results) %>%
  mutate(
    metadata = factor(metadata, levels = fixef),
    feature = gsub("\\.", " ", feature))

# add feed samples to the feature table
mtd_mcs_fd <- filter(mtd, SampleOrigin %in% c("Feed", "Mucosa")) 
taxa_l6_mcs_fd <- select(taxa_l6, rownames(mtd_mcs_fd)) %>%
  mutate(across(everything(), ~.x/sum(.x)))

# filter features showing significant associations with metadata
sigAss_mcs <- filter(fit_mcs_results, qval <= 0.05) %>%
  mutate(N.not.zero = paste0(N.not.zero, "/", N)) %>%
  # NB: taxa names in the maaslin2 outputs may not match those in the feature table!!
  inner_join(rownames_to_column(taxa_l6_mcs_fd, "feature"), by = "feature") %>%
  pivot_longer(
    cols = contains("AqFl1-"), # SampleID starts with "AqFl1-" 
    names_to = "SampleID", 
    values_to = "Abundance") %>%
  inner_join(rownames_to_column(mtd_mcs_fd, "SampleID"), by = "SampleID") %>%
  mutate(
    qval = ifelse(qval < 0.001, "< 0.001", round(qval, 3)), 
    ann_text = paste("q-value:", qval)
  ) 
```

### Heatmap

```{r, fig.width=4, fig.height=8}
hmp_mcs <- plot_heatmap(
  maaslin2_output = fit_mcs_results,
  first_n = n_distinct(sigAss_mcs$feature), 
  cell_value = "qval",
  data_label = "data",
  metadata_label = "metadata",
  plot_title = FALSE,
  legend_title = FALSE,
  legend_title_position = "topcenter", 
  legend_direction = "horizontal", 
  color = c("blue", "grey90", "red"),
  board_line_col = "white",
  colnames_rotate = 45,
  colnames_fontsize = 12,
  rownames_fontsize = 10,
  italize_rownames = TRUE)

draw(
  hmp_mcs, 
  padding = unit(c(2, 2, 2, 20), "mm"), 
  heatmap_legend_side = "top"
 )
```

### Scatter plot

#### Diet

```{r, fig.width=11, fig.height=14}
# filter features showing significant associations with Diet
sigAss_mcs_diet <- sigAss_mcs %>% 
  filter(metadata == "Diet" & SampleOrigin != "Feed") %>%
  group_by(feature, Diet) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(text_extra = paste0("N.not.zero(", Diet, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(text_extra = paste0(text_extra, collapse = "\n")) %>%
  inner_join(filter(sigAss_mcs, metadata == "Diet"), by = "feature") %>%
  mutate(ann_text = paste0(ann_text, "\n", text_extra))

# plotting
filter(sigAss_mcs_diet, SampleOrigin == "Mucosa") %>%
ggplot(aes(x = Diet, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Compartment), position = position_jitter(0.2)) +
  geom_point(
    data = filter(sigAss_mcs_diet, SampleOrigin == "Feed"),
    color = "#666666", size = 3) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_text), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

#### Compartment

```{r, fig.width=11, fig.height=5}
# filter features showing significant associations with Compartment
sigAss_mcs_cmprt <- sigAss_mcs %>% 
  filter(metadata == "Compartment" & SampleOrigin != "Feed")

# add the number of non-zero observations to the text annotation
sigAss_mcs_cmprt <- sigAss_mcs_cmprt %>%
  group_by(feature, Compartment) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(text_extra = paste0("N.not.zero(", Compartment, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(text_extra = paste(text_extra, collapse = "\n")) %>%
  inner_join(sigAss_mcs_cmprt, by = "feature") %>%
  mutate(ann_text = paste0(ann_text, "\n", text_extra))

# plotting
ggplot(sigAss_mcs_cmprt, aes(x = Compartment, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Diet), position = position_jitter(0.2)) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_text), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

### Figure 7

### Figure 7b

Highlight representative taxa enriched in fish fed the REF diet.

```{r, fig.width=6.5, fig.height=5}
# taxa to be shown in the subplot
fig7b_taxa_name <- c(
  "Lactobacillus", "f__Peptostreptococcaceae", "Peptostreptococcus", 
  "Lactococcus", "Leuconostoc", "Pediococcus"
  )

# plot
fig7b <- sigAss_mcs_diet %>% 
  filter(SampleOrigin == "Mucosa" & feature %in% fig7b_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig7b_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Compartment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_mcs_diet %>%
        filter(SampleOrigin == "Feed" & feature %in% fig7b_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig7b_taxa_name)), 
      color = "#666666", size = 3) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_text), size = 3, hjust = 1, vjust = 1) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 1)) +   
    labs(y = "Relative abundance") +
    scale_color_brewer(palette = "Dark2") 

fig7b 
```

### Figure 7c

Highlight representative taxa enriched in fish fed the IM diet.

```{r, fig.width=6.5, fig.height=5}
# taxa to be shown in the subplot
fig7c_taxa_name <- c(
  "Oceanobacillus", "Bacillus", "f__Bacillaceae",  
  "Ornithinibacillus", "Corynebacterium 1", "Enterococcus"
  )

# plot
fig7c <- filter(
  sigAss_mcs_diet, 
  SampleOrigin == "Mucosa" & feature %in% fig7c_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig7c_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Compartment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_mcs_diet %>%
        filter(SampleOrigin == "Feed" & feature %in% fig7c_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig7c_taxa_name)), 
      color = "#666666", size = 3) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_text), size = 3, hjust = 1, vjust = 1) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 1.0)) +   
    labs(y = "Relative abundance") +
    scale_color_brewer(palette = "Dark2") 

fig7c 
```

### Figure 7

Assemble plots.

```{r, fig.width=10, fig.height=12}
# convert the heatmap into a "grob"
fig7_main <- grid.grabExpr(
  draw(
    hmp_mcs, 
    padding = unit(c(2, 2, 2, 8), "mm"), # bottom, left, top and right
    heatmap_legend_side = "top"
   )
  ) 

# assemble subplots
fig7_sub <- fig7b / fig7c + theme(plot.margin = margin(b = 1.5, unit = "cm")) +
  plot_layout(guides = 'collect') 

# heatmap + subplots
wrap_elements(full = fig7_main) + fig7_sub +
  plot_annotation(tag_levels = "a") +
  plot_layout(widths = c(1, 1.6)) 

# export plot
ggsave(here("result", "figure", "Figure 7.tiff"), width = 10, height = 12,
       units = "in", dpi = 300, compression = "lzw")
```

# Session information

```{r}
sessionInfo()
```