---
title: "Multivariable association analysis"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
options(width = 120)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = FALSE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "left",  buttonSide = "left")`

# Getting ready

Set system locale.

```{r}
Sys.setlocale(category = "LC_ALL", locale = "Greek")
```

Load packages.

```{r}
library(here) # A Simpler Way to Find Your Files, CRAN v1.0.1
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN v1.3.0 
library(ggtext) # Improved Text Rendering Support for 'ggplot2', CRAN v0.1.1
library(patchwork) # The Composer of Plots, CRAN v1.1.1  
library(RColorBrewer) # ColorBrewer Palettes, CRAN v1.1-2 
library(PerformanceAnalytics) # Econometric Tools for Performance and Risk Analysis, CRAN v2.0.4
library(MicrobeR) # Handy functions for microbiome analysis in R, [github::jbisanz/MicrobeR] v0.3.2 
library(phyloseq) # Handling and analysis of high-throughput microbiome census data, Bioconductor v1.34.0 
library(Maaslin2) # Maaslin2, Bioconductor v1.4.0
library(parallel) # Support for Parallel computation in R, CRAN v4.0.5
```

Load function.

```{r}
source(here("code/functions/plot_heatmap.R"))
```

Set ggplot global theme.

```{r}
theme_set(
  theme_bw() +
  theme(
    legend.position = "top", 
    strip.text = element_text(face = "italic"),
    axis.title.x = element_markdown())
  )
```

Import data.

```{r}
ps <- readRDS(here("data/intermediate/qiime2R/phyloseq.rds"))
```

# Data preprocessing

## Metadata, OTU table and taxonomy

```{r}
# filter samples
ps <- ps %>%
  subset_samples(SampleOrigin %in% c("Feed", "Digesta", "Mucosa")) %>%
  subset_samples(!(IsTechnicalReplicate == "yes" & PCRBatch == 2))

# extract metadata, OTU table and taxonomy 
mtd <- data.frame(sample_data(ps), check.names = FALSE) %>%
  rownames_to_column("SampleID") %>%
  mutate(FishID = as.character(FishID)) %>%
  # standardize qPCR data for each intestinal compartment
  group_by(Compartment) %>%
  mutate(across(contains("qPCR_"), ~scale(.x))) %>%
  ungroup() %>%
  column_to_rownames("SampleID")
  
otu <- as.data.frame(otu_table(ps)) 

txnm <- tax_table(ps) %>% as("matrix") %>% as.data.frame()
```

## Collapse OTU table at genus-level

Get the lowest taxonomic rank for each OTU.

```{r}
taxa_l6 <- Summarize.Taxa(otu, txnm)$Genus %>%
  rownames_to_column("Taxa") %>%
  separate(
    Taxa, 
    sep = ";", 
    c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")) %>% 
  mutate(
    Class = ifelse(
      is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
      Phylum, 
      Class),
    Order = ifelse(
      is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
      Class, 
      Order),
    Family = ifelse(
      is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
      Order, 
      Family),
    Genus = ifelse(
      is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
      Family,
      Genus)) %>%
  select(-(Kingdom:Family)) %>%
  # remove taxa rank prefix "g__" and square brackets in the taxa names
  mutate(Genus = gsub("g__|\\[|\\]", "", Genus)) %>%
  # merge rows with the same taxa name
  group_by(Genus) %>%
  summarise(across(everything(), sum)) %>%
  column_to_rownames("Genus")
```

# Exploratory data analysis

For the multivariable association analysis, `Diet` and `Compartment` are the main fixed effects we're interested in. We're also interested in potential associations between intestinal microbial clades and gene expression data, which have been published elsewhere ([doi: 10.1016/j.fsi.2018.12.057](https://doi.org/10.1016/j.fsi.2018.12.057)). Based on existing literature, we know that microbiota is related to intestinal immunity and lipid metabolism (digestion, absorption and transport). Before running Maaslin2, let's perform some exploratory data analyses.

### qPCR: immune reponses

Gene expression profile.

```{r, fig.width=6, fig.height=4}
# filter gene expression data related to immune responses 
qpcr_immn <- filter(mtd, SampleOrigin == "Mucosa") %>%
  rename_all(~gsub("qPCR_", "", .x)) %>%
  select(Diet, Compartment, cd3γδ:tgfβ1)
  
# gene expression profile
qpcr_immn %>%
  pivot_longer(cd3γδ:tgfβ1, names_to = "gene", values_to = "rel_expr") %>%
  ggplot(aes(x = Diet, y = rel_expr)) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(aes(color = Compartment), position = position_jitterdodge(0.2)) +
    facet_wrap(~gene, ncol = 4) +
    labs(y = "Standardize expression") +
    theme_bw(base_size = 12) +
    scale_color_brewer(palette = "Dark2") 
```

The expression levels of these immune genes are often correlated. Let's look at their correlations.

```{r, fig.height=8, fig.width=8}
select(qpcr_immn, cd3γδ:tgfβ1) %>% 
  chart.Correlation(mhistogram = TRUE, pch = 19)
```

### qPCR: lipid metabolism

Gene expression profile.

```{r, fig.width=6, fig.height=6}
# filter gene expression data related to lipid metabolism in the intestine
qpcr_lipid  <- filter(mtd, SampleOrigin == "Mucosa") %>%
  rename_all(~gsub("qPCR_", "", .x)) %>%
  select(Diet, Compartment, fabp2b, cd36:cyp51)
  
# gene expression profile
qpcr_lipid %>%
  pivot_longer(fabp2b:cyp51, names_to = "gene", values_to = "rel_expr") %>%
  ggplot(aes(x = Diet, y = rel_expr)) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(aes(color = Compartment), position = position_jitterdodge(0.2)) +
    facet_wrap(~gene, ncol = 3) +
    labs(y = "Standardize expression") +
    theme_bw(base_size = 12) +
    scale_color_brewer(palette = "Dark2") 
```

Correlation plot.

```{r, fig.height=9, fig.width=9}
select(qpcr_lipid, fabp2b:cyp51) %>% 
  chart.Correlation(mhistogram = TRUE, pch = 19)
```

Based on the existing literature and the above exploratory data analysis, we include *cd3γδ*, *foxp3*, *cd36*, *fabp2b*, *apoa4* and *plin2* in the multivariable association analysis in addition to the main effects `Diet` and `Compartment`.

# Multivariable association analysis

Greek letters cause problems when supplied to the `Maaslin2` function. Let's modify the gene name *cd3γδ* to *cd3gd*.

```{r}
mtd <- rename(mtd, qPCR_cd3gd = qPCR_cd3γδ)
```

Define fixed and random effects.

```{r}
# Fixed effects
fixef <- c("Diet", "Compartment", "qPCR_cd3gd", "qPCR_foxp3", 
           "qPCR_cd36", "qPCR_fabp2b", "qPCR_apoa4", "qPCR_plin2") 

# Random effects
ranef <- c("FishID", "Tank")
```

## Digesta

### Filter data

```{r}
# metadata
mtd_dgs <- filter(mtd, SampleOrigin == "Digesta")  

# feature table
taxa_l6_dgs <- select(taxa_l6, rownames(mtd_dgs))
```

### Run MaAsLin2.

```{r, results='hide'}
fit_dgs <- Maaslin2(
  input_data = taxa_l6_dgs, 
  input_metadata = mtd_dgs,
  output = here("data/intermediate/maaslin2/digesta"),
  min_abundance = 0,
  min_prevalence = 0.25,
  max_significance = 0.25,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = fixef,
  random_effects = ranef,
  correction = "BH",
  standardize = TRUE,
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  cores = detectCores()
  )
```

### Tidy MaAsLin2 outputs

Merge maaslin2 outputs, feature table and metadata.

```{r}
# extract maaslin2 output
fit_dgs_results <- filter(fit_dgs$results) %>%
  mutate(
    metadata = gsub("qPCR_", "", metadata),
    metadata = gsub("cd3gd", "cd3γδ", metadata),
    metadata = factor(
      metadata, 
      levels = c("Diet", "Compartment", "cd3γδ", "foxp3", 
                 "cd36", "fabp2b", "apoa4", "plin2")
      )
    ) 

# add feed samples to the feature table
mtd_dgs_fd <- filter(mtd, SampleOrigin %in% c("Feed", "Digesta")) %>%
  rename_at(vars(contains("qPCR_")), ~gsub("qPCR_", "", .x)) %>%
  rename(cd3γδ = cd3gd)

taxa_l6_dgs_fd <- select(taxa_l6, rownames(mtd_dgs_fd))

# filter features showing significant associations with metadata
sigAss_dgs <- filter(fit_dgs_results, qval <= 0.25) %>%
  mutate(
    # fix taxa names in the maaslin2 outputs
    # space (" ") in the original taxa names was replaced with dot(".") in the maaslin2 outputs
    feature = gsub("\\.", " ", feature), 
    N.not.zero = paste0(N.not.zero, "/", N)) %>%
  # NB: taxa names in the maaslin2 outputs may not match those in the feature table!!
  inner_join(rownames_to_column(taxa_l6_dgs_fd, "feature"), by = "feature") %>%
  mutate(across(contains("AqFl1-"), ~.x/sum(.x))) %>%
  pivot_longer(
    cols = contains("AqFl1-"), # SampleID starts with "AqFl1-" 
    names_to = "SampleID", 
    values_to = "Abundance") %>%
  inner_join(rownames_to_column(mtd_dgs_fd, "SampleID"), by = "SampleID") %>%
  mutate(
    qval = ifelse(qval < 0.001, "< 0.001", round(qval, 3)), 
    coef = formatC(coef, format = "e", digits = 1), 
    # text annotation for categorical metadata
    ann_ctg = paste("FDR:", qval),
    # text annotation for continuous metadata
    ann_ctn = paste(
      paste("FDR:", qval), 
      "\n", 
      paste("coefficient:", coef), 
      "\n", 
      paste("N.not.zero:", N.not.zero)
    )
  ) 
```

### Heatmap

Make a customized heatmap using a modified function from the *MaAsLin2* package.

```{r, fig.height=10, fig.width=4}
hmp_dgs <- plot_heatmap(
  output_results = fit_dgs_results,
  first_n = 50, # plot all data: n_distinct(sigAss_dgs$feature),
  cell_value = "qval",
  data_label = "data",
  metadata_label = "metadata",
  plot_title = FALSE,
  legend_title = FALSE,
  legend_title_position = "leftcenter-rot", # topleft, topcenter, leftcenter-rot, lefttop-rot
  color = c("blue", "grey90", "red"),
  board_line_col = "white",
  colnames_rotate = 90,
  colnames_fontsize = 12,
  rownames_fontsize = 10,
  italize_rownames = TRUE)

hmp_dgs
```

### Boxplot: categorical predictors

#### Diet

```{r, fig.width=11, fig.height=38}
# filter features showing significant associations with Diet
sigAss_dgs_diet <- sigAss_dgs %>% 
  filter(metadata == "Diet" & SampleOrigin != "Feed") %>%
  # add the number of non-zero observations in each diet to the text annotation
  group_by(feature, Diet) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Diet, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste0(ann_extra, collapse = "\n")) %>%
  inner_join(filter(sigAss_dgs, metadata == "Diet"), by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
filter(sigAss_dgs_diet, SampleOrigin == "Digesta") %>%
ggplot(aes(x = Diet, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Compartment), position = position_jitter(0.2)) +
  # highlight taxa abundance in the feed samples
  geom_point(
    data = filter(sigAss_dgs_diet, SampleOrigin == "Feed"),
    color = "#666666", size = 3) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

#### Compartment

```{r, fig.width=11, fig.height=12}
# filter features showing significant associations with Compartment
sigAss_dgs_cmprt <- sigAss_dgs %>% 
  filter(metadata == "Compartment" & SampleOrigin != "Feed")

# add the number of non-zero observations to the text annotation
sigAss_dgs_cmprt <- sigAss_dgs_cmprt %>%
  group_by(feature, Compartment) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Compartment, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste(ann_extra, collapse = "\n")) %>%
  inner_join(sigAss_dgs_cmprt, by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
ggplot(sigAss_dgs_cmprt, aes(x = Compartment, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Diet), position = position_jitter(0.2)) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

### Scatter plot: continuous predictors

```{r}
scatter_dgs <- lapply(
  c("cd3γδ", "foxp3", "cd36", "fabp2b", "apoa4", "plin2"), 
  function(x){
    
  var <- ensym(x)
    
  filter(sigAss_dgs, metadata == rlang::as_name(var) & SampleOrigin != "Feed") %>%
  ggplot(aes(x = !!var, y = Abundance)) +
    geom_point(aes(color = Diet), size = 2) +
    geom_smooth(aes(color = Diet), method = "lm", fill = NA) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctn), 
      size = 3, hjust = 1, vjust = 1) +
    facet_wrap(~feature, ncol = 5, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.5)), 
      labels = scales::percent_format(accuracy = 0.1)) +   
    labs(
      x = paste0("Standardized expression of *", rlang::as_name(var), "*"), 
      y = "Relative abundance") +   
    scale_color_brewer(palette = "Dark2") 
  }
)

names(scatter_dgs) <- c("cd3γδ", "foxp3", "cd36", "fabp2b", "apoa4", "plin2")
```

#### qPCR: *cd3γδ*

```{r, fig.width=12, fig.height=6}
scatter_dgs[["cd3γδ"]]
```

#### qPCR: *foxp3*

```{r, fig.width=12, fig.height=6}
scatter_dgs[["foxp3"]]
```

#### qPCR: *cd36*

```{r, fig.width=12, fig.height=6}
scatter_dgs[["cd36"]]
```

#### qPCR: *fabp2b*

```{r, fig.width=12, fig.height=6}
scatter_dgs[["fabp2b"]]
```

#### qPCR: *apoa4*

```{r, fig.width=3, fig.height=4}
scatter_dgs[["apoa4"]]
```

#### qPCR: *plin2*

```{r, fig.width=12, fig.height=6}
scatter_dgs[["plin2"]]
```

## Mucosa

### Filter data

```{r}
# metadata
mtd_mcs <- filter(mtd, SampleOrigin == "Mucosa")  

# feature table
taxa_l6_mcs <- select(taxa_l6, rownames(mtd_mcs))
```

### Run MaAsLin2.

```{r, results='hide'}
fit_mcs <- Maaslin2(
  input_data = taxa_l6_mcs, 
  input_metadata = mtd_mcs,
  output = here("data/intermediate/maaslin2/mucosa"),
  min_abundance = 0,
  min_prevalence = 0.25,
  max_significance = 0.25,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = fixef,
  random_effects = ranef,
  correction = "BH",
  standardize = TRUE,
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  cores = detectCores()
  )
```

### Tidy MaAsLin2 outputs

Merge maaslin2 outputs, feature table and metadata.

```{r}
# extract maaslin2 output
fit_mcs_results <- filter(fit_mcs$results) %>%
  mutate(
    metadata = gsub("qPCR_", "", metadata),
    metadata = gsub("cd3gd", "cd3γδ", metadata),
    metadata = factor(
      metadata, 
      levels = c("Diet", "Compartment", "cd3γδ", "foxp3", 
                 "cd36", "fabp2b", "apoa4", "plin2")
      )
    ) 

# add feed samples to the feature table
mtd_mcs_fd <- filter(mtd, SampleOrigin %in% c("Feed", "Mucosa")) %>%
  rename_at(vars(contains("qPCR_")), ~gsub("qPCR_", "", .x)) %>%
  rename(cd3γδ = cd3gd)

taxa_l6_mcs_fd <- select(taxa_l6, rownames(mtd_mcs_fd))

# filter features showing significant associations with metadata
sigAss_mcs <- filter(fit_mcs_results, qval <= 0.25) %>%
  mutate(
    feature = gsub("\\.", " ", feature), 
    N.not.zero = paste0(N.not.zero, "/", N)) %>%
  inner_join(rownames_to_column(taxa_l6_mcs_fd, "feature"), by = "feature") %>%
  mutate(across(contains("AqFl1-"), ~.x/sum(.x))) %>%
  pivot_longer(
    cols = contains("AqFl1-"), 
    names_to = "SampleID", 
    values_to = "Abundance") %>%
  inner_join(rownames_to_column(mtd_mcs_fd, "SampleID"), by = "SampleID") %>%
  mutate(
    qval = ifelse(qval < 0.001, "< 0.001", round(qval, 3)), 
    coef = formatC(coef, format = "e", digits = 1), 
    ann_ctg = paste("FDR:", qval),
    ann_ctn = paste(
      paste("FDR:", qval), 
      "\n", 
      paste("coefficient:", coef), 
      "\n", 
      paste("N.not.zero:", N.not.zero)
    )
  ) 
```

### Heatmap

```{r, fig.height=8, fig.width=4}
hmp_mcs <- plot_heatmap(
  output_results = fit_mcs_results,
  first_n = n_distinct(sigAss_mcs$feature), # plot all data
  cell_value = "qval",
  data_label = "data",
  metadata_label = "metadata",
  plot_title = FALSE,
  legend_title = FALSE,
  legend_title_position = "leftcenter-rot", 
  color = c("blue", "grey90", "red"),
  board_line_col = "white",
  colnames_rotate = 90,
  colnames_fontsize = 12,
  rownames_fontsize = 10,
  italize_rownames = TRUE)

hmp_mcs
```

### Boxplot: categorical predictors

#### Diet

```{r, fig.width=11, fig.height=12}
# filter features showing significant associations with Diet
sigAss_mcs_diet <- sigAss_mcs %>% 
  filter(metadata == "Diet" & SampleOrigin != "Feed") %>%
  # add the number of non-zero observations in each diet to the text annotation
  group_by(feature, Diet) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Diet, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste0(ann_extra, collapse = "\n")) %>%
  inner_join(filter(sigAss_mcs, metadata == "Diet"), by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
filter(sigAss_mcs_diet, SampleOrigin == "Mucosa") %>%
ggplot(aes(x = Diet, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Compartment), position = position_jitter(0.2)) +
  # highlight taxa abundance in the feed samples
  geom_point(
    data = filter(sigAss_mcs_diet, SampleOrigin == "Feed"),
    color = "#666666", size = 3) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

#### Compartment

```{r, fig.width=11, fig.height=6}
# filter features showing significant associations with Compartment
sigAss_mcs_cmprt <- sigAss_mcs %>% 
  filter(metadata == "Compartment" & SampleOrigin != "Feed")

# add the number of non-zero observations to the text annotation
sigAss_mcs_cmprt <- sigAss_mcs_cmprt %>%
  group_by(feature, Compartment) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Compartment, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste(ann_extra, collapse = "\n")) %>%
  inner_join(sigAss_mcs_cmprt, by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
ggplot(sigAss_mcs_cmprt, aes(x = Compartment, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Diet), position = position_jitter(0.2)) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1, vjust = 1) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)) +   
  labs(y = "Relative abundance") +
  scale_color_brewer(palette = "Dark2") 
```

### Scatter plot: continuous predictors

```{r}
scatter_mcs <- lapply(
  c("cd3γδ", "foxp3", "cd36", "fabp2b", "apoa4", "plin2"), 
  function(x){
    
  var <- ensym(x)
    
  filter(sigAss_mcs, metadata == rlang::as_name(var) & SampleOrigin != "Feed") %>%
  ggplot(aes(x = !!var, y = Abundance)) +
    geom_point(aes(color = Diet), size = 2) +
    geom_smooth(aes(color = Diet), method = "lm", fill = NA) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctn), 
      size = 3, hjust = 1, vjust = 1) +
    facet_wrap(~feature, ncol = 5, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.5)), 
      labels = scales::percent_format(accuracy = 0.1)) +   
    labs(
      x = paste0("Standardized expression of *", rlang::as_name(var), "*"), 
      y = "Relative abundance") +   
    scale_color_brewer(palette = "Dark2") 
  }
)

names(scatter_mcs) <- c("cd3γδ", "foxp3", "cd36", "fabp2b", "apoa4", "plin2")
```

#### qPCR: *cd3γδ*

No significant associations identified.

#### qPCR: *foxp3*

```{r, fig.width=4.8, fig.height=3}
scatter_mcs[["foxp3"]]
```

#### qPCR: *cd36*

No significant associations identified.

#### qPCR: *fabp2b*

```{r, fig.width=3, fig.height=4}
scatter_mcs[["fabp2b"]]
```

#### qPCR: *apoa4*

```{r, fig.width=3, fig.height=4}
scatter_mcs[["apoa4"]]
```

#### qPCR: *plin2*

```{r, fig.width=12, fig.height=6}
scatter_mcs[["plin2"]]
```

# Session information

```{r}
sessionInfo()
```
