---
title: "Multivariable association analysis"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
options(width = 120)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = TRUE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "right",  buttonSide = "right")`

# Getting ready

Set system locale.

```{r, results='hide'}
Sys.setlocale(category = "LC_ALL", locale = "Greek")
```

Load packages.

```{r}
library(here) # A Simpler Way to Find Your Files, CRAN 
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN
library(ggtext) # Improved Text Rendering Support for 'ggplot2', CRAN 
library(patchwork) # The Composer of Plots, CRAN   
library(RColorBrewer) # ColorBrewer Palettes, CRAN 
library(PerformanceAnalytics) # Econometric Tools for Performance and Risk Analysis, CRAN 
library(MicrobeR) # Handy functions for microbiome analysis in R, [github::jbisanz/MicrobeR]  
library(phyloseq) # Handling and analysis of high-throughput microbiome census data, Bioconductor
library(Maaslin2) # Maaslin2, Bioconductor 

# Set seed
set.seed(1910)
```

Load function.

```{r}
source(here("code/functions/plot_heatmap.R"))
```

Set ggplot global theme.

```{r}
theme_set(
  theme_bw() +
  theme(
    legend.position = "top", 
    strip.text = element_text(face = "italic"),
    axis.title.x = element_markdown()
    )
  )
```

Import data.

```{r}
ps <- readRDS(here("data/intermediate/qiime2R/phyloseq.rds"))
```

# Data preprocessing

## Metadata, feature table and taxonomy

```{r}
# filter samples
ps <- ps %>%
  subset_samples(Source %in% c("Feed", "Digesta", "Mucosa")) %>%
  subset_samples(!(IsTechnicalReplicate == "yes" & PCRBatch == 2))

# extract metadata 
mtd <- data.frame(sample_data(ps), check.names = FALSE) %>%
  rownames_to_column("SampleID") %>%
  mutate(FishID = as.character(FishID)) %>%
  # standardize qPCR data for each intestinal Segment
  group_by(Segment) %>%
  mutate(across(contains("qPCR_"), ~scale(.x))) %>%
  ungroup() %>%
  # remove gene name prefix
  rename_at(vars(contains("qPCR_")), ~gsub("qPCR_", "", .x)) %>%
  column_to_rownames("SampleID")
  
# extract count table 
count_tbl <- as.data.frame(otu_table(ps)) 

# extract taxonomy table 
txnm_tbl <- tax_table(ps) %>% as("matrix") %>% as.data.frame()
```

## Collapse feature table at genus-level

```{r}
taxa_l6 <- Summarize.Taxa(count_tbl, txnm_tbl)$Genus %>%
  rownames_to_column("Taxa") %>%
  separate(
    Taxa, 
    sep = ";", 
    c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")) %>% 
  mutate(
    Class = ifelse(
      is.na(Class)|Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), 
      Phylum, 
      Class),
    Order = ifelse(
      is.na(Order)|Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), 
      Class, 
      Order),
    Family = ifelse(
      is.na(Family)|Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), 
      Order, 
      Family),
    Genus = ifelse(
      is.na(Genus)|Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), 
      Family,
      Genus)) %>%
  select(-(Kingdom:Family)) %>%
  # remove taxa rank prefix "g__" and square brackets in the taxa names
  mutate(Genus = gsub("g__|\\[|\\]", "", Genus)) %>%
  # merge rows with the same taxa name
  group_by(Genus) %>%
  summarise(across(everything(), sum)) %>%
  column_to_rownames("Genus")
```

# Exploratory data analysis

For the multivariable association analysis, `Diet` and `Segment` are the main fixed effects we're interested in. We're also interested in potential associations between intestinal microbial clades and gene expression data, which have been published elsewhere ([(Li et al., 2019](https://doi.org/10.1016/j.fsi.2018.12.057)). Based on existing literature, we know that microbiota is related to intestinal immunity and lipid metabolism (digestion, absorption and transport). Before running Maaslin2, let's perform some exploratory data analyses.

### qPCR: immune reponses

Gene expression profile.

```{r, fig.width=6, fig.height=4}
# filter gene expression data related to immune responses 
qpcr_immn <- filter(mtd, Source == "Mucosa") %>%
  select(Diet, Segment, cd3γδ:tgfβ1)
  
# gene expression profile
qpcr_immn %>%
  pivot_longer(cd3γδ:tgfβ1, names_to = "gene", values_to = "rel_expr") %>%
  ggplot(aes(x = Diet, y = rel_expr)) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(aes(color = Segment), position = position_jitterdodge(0.2)) +
    facet_wrap(~gene, ncol = 4) +
    labs(y = "Standardize expression") +
    theme_bw(base_size = 12) +
    scale_color_brewer(palette = "Dark2") 
```

The expression levels of these immune genes are often correlated. Let's look at their correlations.

```{r, fig.height=8, fig.width=8}
select(qpcr_immn, cd3γδ:tgfβ1) %>% 
  chart.Correlation(mhistogram = TRUE, pch = 19)
```

### qPCR: lipid metabolism

Gene expression profile.

```{r, fig.width=6, fig.height=6}
# filter gene expression data related to lipid metabolism in the intestine
qpcr_lipid  <- filter(mtd, Source == "Mucosa") %>%
  select(Diet, Segment, fabp2b, cd36:cyp51)
  
# gene expression profile
qpcr_lipid %>%
  pivot_longer(fabp2b:cyp51, names_to = "gene", values_to = "rel_expr") %>%
  ggplot(aes(x = Diet, y = rel_expr)) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(aes(color = Segment), position = position_jitterdodge(0.2)) +
    facet_wrap(~gene, ncol = 3) +
    labs(y = "Standardize expression") +
    theme_bw(base_size = 12) +
    scale_color_brewer(palette = "Dark2") 
```

Correlation plot.

```{r, fig.height=9, fig.width=9}
select(qpcr_lipid, fabp2b:cyp51) %>% 
  chart.Correlation(mhistogram = TRUE, pch = 19)
```

Based on the existing literature and the above exploratory data analysis, we include *foxp3* and *plin2* in the multivariable association analysis in addition to the main effects `Diet` and `Segment`.

# Multivariable association analysis

Define fixed and random effects.

```{r}
# fixed effects
fixef <- c("Diet", "Segment", "foxp3", "plin2") 

# random effects
ranef <- c("FishID", "Tank")
```

## Digesta

### Filter data

```{r}
# metadata
mtd_dgs <- filter(mtd, Source == "Digesta")  

# feature table
taxa_l6_dgs <- select(taxa_l6, rownames(mtd_dgs))
```

### Run MaAsLin2.

```{r, results='hide'}
fit_dgs <- Maaslin2(
  input_data = taxa_l6_dgs, 
  input_metadata = mtd_dgs,
  output = here("data/intermediate/maaslin2/digesta"),
  min_abundance = 0,
  min_prevalence = 0.25,
  max_significance = 0.1,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = fixef,
  random_effects = ranef,
  correction = "BH",
  standardize = FALSE, # already did during the data preprocessing 
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  cores = parallel::detectCores()
  )
```

### Tidy MaAsLin2 outputs

Merge maaslin2 outputs, feature table and metadata.

```{r}
# extract maaslin2 output
fit_dgs_results <- fit_dgs$results %>%
  mutate(
    metadata = factor(metadata, levels = fixef),
    # fix taxa names in the MaAsLin2 outputs: space in taxa names was replaced with dot
    feature = gsub("\\.", " ", feature))

# add feed samples to the feature table
mtd_dgs_fd <- filter(mtd, Source %in% c("Feed", "Digesta"))
taxa_l6_dgs_fd <- select(taxa_l6, rownames(mtd_dgs_fd)) %>%
  mutate(across(everything(), ~.x/sum(.x)))

# filter features showing significant associations with metadata
sigAss_dgs <- filter(fit_dgs_results, qval <= 0.1) %>%
  mutate(N.not.zero = paste0(N.not.zero, "/", N)) %>%
  # NB: taxa names in the maaslin2 outputs may not match those in the feature table!!
  inner_join(rownames_to_column(taxa_l6_dgs_fd, "feature"), by = "feature") %>%
  pivot_longer(
    cols = contains("AqFl1-"), # SampleID starts with "AqFl1-" 
    names_to = "SampleID", 
    values_to = "Abundance"
    ) %>%
  inner_join(rownames_to_column(mtd_dgs_fd, "SampleID"), by = "SampleID") %>%
  mutate(
    qval = ifelse(qval < 0.001, "< 0.001", round(qval, 3)), 
    coef = formatC(coef, format = "e", digits = 1), 
    # text annotation for categorical metadata
    ann_ctg = paste("FDR:", qval),
    # text annotation for continuous metadata
    ann_ctn = paste(
      paste("FDR:", qval), 
      "\n", 
      paste("coefficient:", coef), 
      "\n", 
      paste("N.not.zero:", N.not.zero)
    )
  ) 
```

### Heatmap

In total, 89 features showed significant associations with `Diet`, which are too many to be displayed on the heatmap. Here we filter these features to keep those with a q-value < 1e-4, which results in 63 features in total to be displayed.

```{r, fig.width=4.5, fig.height=10}
# filter features
feature_to_display <- fit_dgs_results %>%
  filter(qval <= 0.1) %>%
  filter(!(metadata == "Diet" & qval >= 1e-4)) %>%
  pull(feature) %>%
  unique()

# make heatmap  
hmp_dgs <- plot_heatmap(
  maaslin2_output = filter(fit_dgs_results, feature %in% feature_to_display),
  qval_cutoff = 0.1,
  first_n = length(feature_to_display), # plot all data: n_distinct(sigAss_dgs$feature)
  cell_value = "qval",
  data_label = "data",
  metadata_label = "metadata",
  plot_title = FALSE,
  legend_title = FALSE,
  legend_title_position = "topcenter", # topcenter, topleft, leftcenter-rot, lefttop-rot
  legend_direction = "horizontal", # "vertical", "horizontal"
  color = c("blue", "grey90", "red"),
  board_line_col = "white",
  colnames_rotate = 50,
  colnames_fontsize = 10,
  rownames_fontsize = 10,
  italize_rownames = TRUE)

draw(
  hmp_dgs, 
  padding = unit(c(2, 2, 2, 20), "mm"), # bottom, left, top and right
  heatmap_legend_side = "top"
 )
```

### Boxplot: categorical predictors

#### Diet

```{r, fig.width=11, fig.height=36}
# filter features showing significant associations with Diet
sigAss_dgs_diet <- sigAss_dgs %>% 
  filter(metadata == "Diet" & Source != "Feed") %>%
  # add the number of non-zero observations in each diet to the text annotation
  group_by(feature, Diet) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Diet, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste0(ann_extra, collapse = "\n")) %>%
  inner_join(filter(sigAss_dgs, metadata == "Diet"), by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
filter(sigAss_dgs_diet, Source == "Digesta") %>%
ggplot(aes(x = Diet, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Segment), position = position_jitter(0.2)) +
  # highlight taxa abundance in the feed samples (large grey dots)
  geom_point(
    data = filter(sigAss_dgs_diet, Source == "Feed"),
    color = "#666666", size = 3
    ) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1.05, vjust = 1.05
    ) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)
    ) +   
  labs(y = "Relative abundance") +
  scale_color_manual(values = brewer.pal(4, "Dark2")[c(1:2)]) 
```

#### Segment

```{r, fig.width=11, fig.height=11}
# filter features showing significant associations with Segment
sigAss_dgs_seg <- sigAss_dgs %>% 
  filter(metadata == "Segment" & Source != "Feed")

# add the number of non-zero observations to the text annotation
sigAss_dgs_seg <- sigAss_dgs_seg %>%
  group_by(feature, Segment) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Segment, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste(ann_extra, collapse = "\n")) %>%
  inner_join(sigAss_dgs_seg, by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
ggplot(sigAss_dgs_seg, aes(x = Segment, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Diet), position = position_jitter(0.2)) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1.05, vjust = 1.05
    ) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)
    ) +   
  labs(y = "Relative abundance") +
  scale_color_manual(values = brewer.pal(4, "Dark2")[c(3:4)])  
```

### Scatter plot: continuous predictors

```{r}
scatter_dgs <- lapply(
  fixef[-c(1:2)], 
  function(x){
    
  var <- ensym(x)
    
  filter(sigAss_dgs, metadata == rlang::as_name(var) & Source != "Feed") %>%
  ggplot(aes(x = !!var, y = log(Abundance))) +
    geom_point(aes(color = Diet), size = 2) +
    geom_smooth(color = "#666666", method = "lm", fill = NA) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctn), 
      size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 5, scales = "free_y") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.5))) +   
    labs(
      x = paste0("Standardized expression of *", rlang::as_name(var), "*"), 
      y = "Relative abundance (log scale)"
      ) +   
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(3:4)])  
  }
)

names(scatter_dgs) <- fixef[-c(1:2)]
```

#### qPCR: *foxp3*

```{r, fig.width=10, fig.height=4}
scatter_dgs[["foxp3"]]
```

#### qPCR: *plin2*

```{r, fig.width=10, fig.height=4}
scatter_dgs[["plin2"]]
```

## Figure 5

### Figure 5b

Highlight representative taxa enriched in fish fed the REF diet.

```{r, fig.width=6.5, fig.height=5}
# taxa to be shown in the subplot
fig6b_taxa_name <- c(
  "Lactobacillus", "f__Peptostreptococcaceae", "Peptostreptococcus", 
  "Lactococcus", "Leuconostoc", "Pediococcus"
  )

# plot
fig6b <- sigAss_dgs_diet %>% 
  filter(Source == "Digesta" & feature %in% fig6b_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig6b_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Segment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_dgs_diet %>%
        filter(Source == "Feed" & feature %in% fig6b_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig6b_taxa_name)), 
      color = "#666666", size = 3
      ) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctg), size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 0.1, suffix = "")
      ) +   
    labs(y = "Relative abundance (%)") +
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(1:2)]) 

fig6b 
```

### Figure 5c

Highlight representative taxa enriched in fish fed the IM diet.

```{r, fig.width=6.5, fig.height=5} 
# taxa to be shown in the subplot
fig6c_taxa_name <- c(
  "Oceanobacillus", "Bacillus", "f__Bacillaceae",  
  "Ornithinibacillus", "Lysinibacillus", "Gracilibacillus" 
  )

# plot
fig6c <- filter(
  sigAss_dgs_diet, 
  Source == "Digesta" & feature %in% fig6c_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig6c_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Segment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_dgs_diet %>%
        filter(Source == "Feed" & feature %in% fig6c_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig6c_taxa_name)), 
      color = "#666666", size = 3
      ) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctg), size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 0.1, suffix = "")
      ) +   
    labs(y = "Relative abundance (%)") +
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(1:2)])

fig6c 
```

### Figure 5

Assemble plots.

```{r, fig.width=10, fig.height=12}
# convert the heatmap into a "grob"
fig6_main <- grid.grabExpr(
  draw(
    hmp_dgs, 
    padding = unit(c(2, 2, 2, 8), "mm"), # bottom, left, top and right
    heatmap_legend_side = "top"
   )
  ) 

# assemble subplots
fig6_sub <- fig6b / fig6c + theme(plot.margin = margin(b = 0.75, unit = "cm")) +
  plot_layout(guides = 'collect') 

# heatmap + subplots
wrap_elements(full = fig6_main) + fig6_sub +
  plot_annotation(tag_levels = "a") +
  plot_layout(widths = c(1, 1.6)) 

# export plot
ggsave(here("result", "figure", "Figure 5.tiff"), width = 10, height = 12,
       units = "in", dpi = 300, compression = "lzw")
```

## Mucosa

### Filter data

```{r}
# metadata
mtd_mcs <- filter(mtd, Source == "Mucosa")  

# feature table
taxa_l6_mcs <- select(taxa_l6, rownames(mtd_mcs))
```

### Run MaAsLin2.

```{r, results='hide'}
fit_mcs <- Maaslin2(
  input_data = taxa_l6_mcs, 
  input_metadata = mtd_mcs,
  output = here("data/intermediate/maaslin2/mucosa"),
  min_abundance = 0,
  min_prevalence = 0.25,
  max_significance = 0.1,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = fixef,
  random_effects = ranef,
  correction = "BH",
  standardize = FALSE, 
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  cores = parallel::detectCores()
  )
```

### Tidy MaAsLin2 outputs

```{r}
# extract maaslin2 output
fit_mcs_results <- fit_mcs$results %>%
  mutate(
    metadata = factor(metadata, levels = fixef),
    feature = gsub("\\.", " ", feature))

# add feed samples to the feature table
mtd_mcs_fd <- filter(mtd, Source %in% c("Feed", "Mucosa")) 
taxa_l6_mcs_fd <- select(taxa_l6, rownames(mtd_mcs_fd)) %>%
  mutate(across(everything(), ~.x/sum(.x)))

# filter features showing significant associations with metadata
sigAss_mcs <- filter(fit_mcs_results, qval <= 0.1) %>%
  mutate(
    feature = gsub("\\.", " ", feature), 
    N.not.zero = paste0(N.not.zero, "/", N)
    ) %>%
  inner_join(rownames_to_column(taxa_l6_mcs_fd, "feature"), by = "feature") %>%
  pivot_longer(
    cols = contains("AqFl1-"), 
    names_to = "SampleID", 
    values_to = "Abundance"
    ) %>%
  inner_join(rownames_to_column(mtd_mcs_fd, "SampleID"), by = "SampleID") %>%
  mutate(
    qval = ifelse(qval < 0.001, "< 0.001", round(qval, 3)), 
    coef = formatC(coef, format = "e", digits = 1), 
    ann_ctg = paste("FDR:", qval),
    ann_ctn = paste(
      paste("FDR:", qval), 
      "\n", 
      paste("coefficient:", coef), 
      "\n", 
      paste("N.not.zero:", N.not.zero)
    )
  ) 
```

### Heatmap

```{r, fig.height=8, fig.width=4}
hmp_mcs <- plot_heatmap(
  maaslin2_output = fit_mcs_results,
  qval_cutoff = 0.1,
  first_n = n_distinct(sigAss_mcs$feature), # plot all data
  cell_value = "qval",
  data_label = "data",
  metadata_label = "metadata",
  plot_title = FALSE,
  legend_title = FALSE,
  legend_title_position = "topcenter",
  legend_direction = "horizontal", 
  color = c("blue", "grey90", "red"),
  board_line_col = "white",
  colnames_rotate = 50,
  colnames_fontsize = 10,
  rownames_fontsize = 10,
  italize_rownames = TRUE)

draw(
  hmp_mcs, 
  padding = unit(c(2, 2, 2, 20), "mm"), 
  heatmap_legend_side = "top"
 )
```

### Boxplot: categorical predictors

#### Diet

```{r, fig.width=11, fig.height=14}
# filter features showing significant associations with Diet
sigAss_mcs_diet <- sigAss_mcs %>% 
  filter(metadata == "Diet" & Source != "Feed") %>%
  group_by(feature, Diet) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Diet, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste0(ann_extra, collapse = "\n")) %>%
  inner_join(filter(sigAss_mcs, metadata == "Diet"), by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
filter(sigAss_mcs_diet, Source == "Mucosa") %>%
ggplot(aes(x = Diet, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Segment), position = position_jitter(0.2)) +
  geom_point(
    data = filter(sigAss_mcs_diet, Source == "Feed"),
    color = "#666666", size = 3
    ) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1.05, vjust = 1.05
    ) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)
    ) +   
  labs(y = "Relative abundance") +
  scale_color_manual(values = brewer.pal(4, "Dark2")[c(1:2)])   
```

#### Segment

```{r, fig.width=11, fig.height=6.5}
# filter features showing significant associations with Segment
sigAss_mcs_seg <- sigAss_mcs %>% 
  filter(metadata == "Segment" & Source != "Feed")

# add the number of non-zero observations to the text annotation
sigAss_mcs_seg <- sigAss_mcs_seg %>%
  group_by(feature, Segment) %>%
  summarize(N = n(), N_notzero = paste0(sum(Abundance != 0), "/", N)) %>%
  mutate(ann_extra = paste0("N.not.zero(", Segment, "): ", N_notzero)) %>%
  group_by(feature) %>%
  summarize(ann_extra = paste(ann_extra, collapse = "\n")) %>%
  inner_join(sigAss_mcs_seg, by = "feature") %>%
  mutate(ann_ctg = paste0(ann_ctg, "\n", ann_extra))

# plotting
ggplot(sigAss_mcs_seg, aes(x = Segment, y = Abundance)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(aes(color = Diet), position = position_jitter(0.2)) +
  geom_text(
    aes(x = Inf, y = Inf, label = ann_ctg), 
    size = 3, hjust = 1.05, vjust = 1.05
    ) +
  facet_wrap(~feature, ncol = 5, scales = "free_y") +
  scale_y_continuous(
    limits = c(0, NA),expand = expansion(mult = c(0, 0.6)), 
    labels = scales::percent_format(accuracy = 0.1)
    ) +   
  labs(y = "Relative abundance") +
  scale_color_manual(values = brewer.pal(4, "Dark2")[c(3:4)])  
```

### Scatter plot: continuous predictors

```{r}
scatter_mcs <- lapply(
  fixef[-c(1:2)], 
  function(x){
    
  var <- ensym(x)
    
  filter(sigAss_mcs, metadata == rlang::as_name(var) & Source != "Feed") %>%
  ggplot(aes(x = !!var, y = log(Abundance))) +
    geom_point(aes(color = Diet), size = 2) +
    geom_smooth(color = "#666666", method = "lm", fill = NA) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctn), 
      size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 5, scales = "free_y") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.5))) +   
    labs(
      x = paste0("Standardized expression of *", rlang::as_name(var), "*"), 
      y = "Relative abundance (log scale)"
      ) +   
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(3:4)])  
  }
)

names(scatter_mcs) <- fixef[-c(1:2)]
```

#### qPCR: *foxp3*

```{r, fig.width=7.5, fig.height=4}
scatter_mcs[["foxp3"]]
```

#### qPCR: *plin2*

```{r, fig.width=12.5, fig.height=4}
scatter_mcs[["plin2"]] + facet_wrap(~feature, ncol = 5, scales = "free_y")
```

## Figure 6

### Figure 6b

Highlight representative taxa enriched in fish fed the REF diet.

```{r, fig.width=6.5, fig.height=3}
# taxa to be shown in the subplot
fig7b_taxa_name <- c(
  "Lactobacillus", "f__Peptostreptococcaceae", "Peptostreptococcus"
  )

# plot
fig7b <- sigAss_mcs_diet %>% 
  filter(Source == "Mucosa" & feature %in% fig7b_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig7b_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Segment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_mcs_diet %>%
        filter(Source == "Feed" & feature %in% fig7b_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig7b_taxa_name)), 
      color = "#666666", size = 3
      ) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctg), size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 0.1, suffix = "")
      ) +   
    labs(y = "Relative abundance (%)") +
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(1:2)])

fig7b 
```

### Figure 6c

Highlight representative taxa enriched in fish fed the IM diet.

```{r, fig.width=6.5, fig.height=5}
# taxa to be shown in the subplot
fig7c_taxa_name <- c(
  "Oceanobacillus", "Bacillus", "f__Bacillaceae",  
  "Ornithinibacillus", "Lysinibacillus", "Gracilibacillus" 
  )

# plot
fig7c <- filter(
  sigAss_mcs_diet, 
  Source == "Mucosa" & feature %in% fig7c_taxa_name) %>%
  mutate(feature = factor(feature, levels = fig7c_taxa_name)) %>%
  ggplot(aes(x = Diet, y = Abundance)) +
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    geom_jitter(aes(color = Segment), position = position_jitter(0.2))  +
    geom_point(
      data = sigAss_mcs_diet %>%
        filter(Source == "Feed" & feature %in% fig7c_taxa_name) %>%
        mutate(feature = factor(feature, levels = fig7c_taxa_name)), 
      color = "#666666", size = 3
      ) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctg), size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 3, scales = "free_y") +
    scale_y_continuous(
      limits = c(0, NA), expand = expansion(mult = c(0, 0.6)), 
      labels = scales::percent_format(accuracy = 0.1, suffix = "")
      ) +   
    labs(y = "Relative abundance (%)") +
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(1:2)])

fig7c 
```

### Figure 6d

Highlight the positive association between *Paenibacillus* and *foxp3*.

```{r, fig.width=3, fig.height=4}
fig7d <- sigAss_mcs %>%
  filter(metadata == "foxp3" & Source != "Feed") %>%
  filter(feature == "Paenibacillus") %>%
  ggplot(aes(x = foxp3, y = log(Abundance))) +
    geom_point(aes(color = Diet), size = 2) +
    geom_smooth(color = "#666666", method = "lm", fill = NA) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctn), 
      size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 5, scales = "free_y") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.4))) +     
    labs(
      x = "*foxp3* (standardized expression)", 
      y = "Relative abundance (log scale)"
      ) +   
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(3:4)])

fig7d
```

### Figure 6e

Highlight the negative association between *Enterococcus* and *plin2*.

```{r, fig.width=3, fig.height=4}
fig7e <- sigAss_mcs %>%
  filter(metadata == "plin2" & Source != "Feed") %>%
  filter(feature == "Enterococcus") %>%
  ggplot(aes(x = plin2, y = log(Abundance))) +
    geom_point(aes(color = Diet), size = 2) +
    geom_smooth(color = "#666666", method = "lm", fill = NA) +
    geom_text(
      aes(x = Inf, y = Inf, label = ann_ctn), 
      size = 3, hjust = 1.05, vjust = 1.05
      ) +
    facet_wrap(~feature, ncol = 5, scales = "free_y") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.4))) +   
    labs(
      x = "*plin2* (standardized expression)", 
      y = "Relative abundance (log scale)"
      ) +   
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(3:4)])

fig7e
```

### Figure 6

Assemble plots.

```{r, fig.width=10, fig.height=12}
# convert the heatmap into a "grob"
fig7_main <- grid.grabExpr(
  draw(
    hmp_mcs, 
    padding = unit(c(6, 2, 2, 8), "mm"), # bottom, left, top and right
    heatmap_legend_side = "top"
   )
  ) 

# assemble subplots
fig7bc <- fig7b / fig7c +
  plot_layout(heights = c(1, 2), guides = 'collect') 

fig7de <- fig7d + fig7e * theme(plot.margin = margin(b = 0, unit = "cm")) +
  plot_layout(guides = 'collect') &
  theme(legend.position = "bottom")

fig7_sub <- fig7bc / fig7de +
  plot_layout(nrow = 3, heights = c(1, 2, 1.2)) 

# heatmap + subplots
wrap_elements(full = fig7_main) + fig7_sub +
  plot_annotation(tag_levels = "a") +
  plot_layout(widths = c(1, 1.6)) 

# export plot
ggsave(here("result", "figure", "Figure 6.tiff"), width = 10, height = 12,
       units = "in", dpi = 300, compression = "lzw")
```

# Session information

```{r}
sessionInfo()
```
